<!doctype html>
<html class="no-js" lang="en" data-content_root="../">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<link rel="index" title="Index" href="../genindex.html" /><link rel="search" title="Search" href="../search.html" /><link rel="next" title="pertpy.data.adamson_2016_pilot" href="data/pertpy.data.adamson_2016_pilot.html" /><link rel="prev" title="Installation" href="../installation.html" />

    <!-- Generated with Sphinx 7.2.3 and Furo 2023.08.19 -->
        <title>Usage - pertpy</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=80d5e7a1" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo.css?v=135e06be" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../_static/sg_gallery.css?v=61a4c737" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.1e8bd061cd6da7fc9cf755528e8ffc24.min.css?v=0a3b3ea7" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo-extensions.css?v=36a5483c" />
    <link rel="stylesheet" type="text/css" href="../_static/css/override.css" />
    <link rel="stylesheet" type="text/css" href="../_static/css/sphinx_gallery.css?v=4d525498" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  --color-brand-primary: #003262;
  --color-brand-content: #003262per;
  --admonition-font-size: var(--font-size-normal);
  --admonition-title-font-size: var(--font-size-normal);
  --code-font-size: var(--font-size--small);
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../index.html"><div class="brand">pertpy</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../index.html">
  
  
  
</a><form class="sidebar-search-container" method="get" action="../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1 current has-children current-page"><a class="current reference internal" href="#">Usage</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle navigation of Usage</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.adamson_2016_pilot.html">pertpy.data.adamson_2016_pilot</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.adamson_2016_upr_epistasis.html">pertpy.data.adamson_2016_upr_epistasis</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.adamson_2016_upr_perturb_seq.html">pertpy.data.adamson_2016_upr_perturb_seq</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.aissa_2021.html">pertpy.data.aissa_2021</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.bhattacherjee.html">pertpy.data.bhattacherjee</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.burczynski_crohn.html">pertpy.data.burczynski_crohn</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.chang_2021.html">pertpy.data.chang_2021</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.datlinger_2017.html">pertpy.data.datlinger_2017</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.datlinger_2021.html">pertpy.data.datlinger_2021</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.dialogue_example.html">pertpy.data.dialogue_example</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.dixit_2016.html">pertpy.data.dixit_2016</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.dixit_2016_raw.html">pertpy.data.dixit_2016_raw</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.frangieh_2021.html">pertpy.data.frangieh_2021</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.frangieh_2021_protein.html">pertpy.data.frangieh_2021_protein</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.frangieh_2021_raw.html">pertpy.data.frangieh_2021_raw</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.frangieh_2021_rna.html">pertpy.data.frangieh_2021_rna</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.gasperini_2019_atscale.html">pertpy.data.gasperini_2019_atscale</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.gasperini_2019_highmoi.html">pertpy.data.gasperini_2019_highmoi</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.gasperini_2019_lowmoi.html">pertpy.data.gasperini_2019_lowmoi</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.gehring_2019.html">pertpy.data.gehring_2019</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.haber_2017_regions.html">pertpy.data.haber_2017_regions</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.kang_2018.html">pertpy.data.kang_2018</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.mcfarland_2020.html">pertpy.data.mcfarland_2020</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.norman_2019.html">pertpy.data.norman_2019</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.norman_2019_raw.html">pertpy.data.norman_2019_raw</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.papalexi_2021.html">pertpy.data.papalexi_2021</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.replogle_2022_k562_essential.html">pertpy.data.replogle_2022_k562_essential</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.replogle_2022_k562_gwps.html">pertpy.data.replogle_2022_k562_gwps</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.replogle_2022_rpe1.html">pertpy.data.replogle_2022_rpe1</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.sc_sim_augur.html">pertpy.data.sc_sim_augur</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.schiebinger_2019_16day.html">pertpy.data.schiebinger_2019_16day</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.schiebinger_2019_18day.html">pertpy.data.schiebinger_2019_18day</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.schraivogel_2020_tap_screen_chr8.html">pertpy.data.schraivogel_2020_tap_screen_chr8</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.schraivogel_2020_tap_screen_chr11.html">pertpy.data.schraivogel_2020_tap_screen_chr11</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.sciplex3_raw.html">pertpy.data.sciplex3_raw</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.shifrut_2018.html">pertpy.data.shifrut_2018</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.smillie.html">pertpy.data.smillie</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.srivatsan_2020_sciplex2.html">pertpy.data.srivatsan_2020_sciplex2</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.srivatsan_2020_sciplex3.html">pertpy.data.srivatsan_2020_sciplex3</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.srivatsan_2020_sciplex4.html">pertpy.data.srivatsan_2020_sciplex4</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.stephenson_2021_subsampled.html">pertpy.data.stephenson_2021_subsampled</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.tian_2019_day7neuron.html">pertpy.data.tian_2019_day7neuron</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.tian_2019_ipsc.html">pertpy.data.tian_2019_ipsc</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.tian_2021_crispra.html">pertpy.data.tian_2021_crispra</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.tian_2021_crispri.html">pertpy.data.tian_2021_crispri</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.weinreb_2020.html">pertpy.data.weinreb_2020</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.xie_2017.html">pertpy.data.xie_2017</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.zhao_2021.html">pertpy.data.zhao_2021</a></li>
<li class="toctree-l2"><a class="reference internal" href="preprocessing/pertpy.preprocessing.GuideAssignment.html">pertpy.preprocessing.GuideAssignment</a></li>
<li class="toctree-l2"><a class="reference internal" href="tools/pertpy.tools.Mixscape.html">pertpy.tools.Mixscape</a></li>
<li class="toctree-l2"><a class="reference internal" href="tools/pertpy.tools.Milo.html">pertpy.tools.Milo</a></li>
<li class="toctree-l2"><a class="reference internal" href="tools/pertpy.tools.Sccoda.html">pertpy.tools.Sccoda</a></li>
<li class="toctree-l2"><a class="reference internal" href="tools/pertpy.tools.Tasccoda.html">pertpy.tools.Tasccoda</a></li>
<li class="toctree-l2"><a class="reference internal" href="tools/pertpy.tools.Dialogue.html">pertpy.tools.Dialogue</a></li>
<li class="toctree-l2"><a class="reference internal" href="tools/pertpy.tools.Distance.html">pertpy.tools.Distance</a></li>
<li class="toctree-l2"><a class="reference internal" href="tools/pertpy.tools.DistanceTest.html">pertpy.tools.DistanceTest</a></li>
<li class="toctree-l2"><a class="reference internal" href="tools/pertpy.tools.CellLineMetaData.html">pertpy.tools.CellLineMetaData</a></li>
<li class="toctree-l2"><a class="reference internal" href="tools/pertpy.tools.Augur.html">pertpy.tools.Augur</a></li>
<li class="toctree-l2"><a class="reference internal" href="tools/pertpy.tools.SCGEN.html">pertpy.tools.SCGEN</a></li>
<li class="toctree-l2"><a class="reference internal" href="tools/pertpy.tools.DiscriminatorClassifierSpace.html">pertpy.tools.DiscriminatorClassifierSpace</a></li>
<li class="toctree-l2"><a class="reference internal" href="tools/pertpy.tools.CentroidSpace.html">pertpy.tools.CentroidSpace</a></li>
<li class="toctree-l2"><a class="reference internal" href="tools/pertpy.tools.DBSCANSpace.html">pertpy.tools.DBSCANSpace</a></li>
<li class="toctree-l2"><a class="reference internal" href="tools/pertpy.tools.KMeansSpace.html">pertpy.tools.KMeansSpace</a></li>
<li class="toctree-l2"><a class="reference internal" href="tools/pertpy.tools.PseudobulkSpace.html">pertpy.tools.PseudobulkSpace</a></li>
<li class="toctree-l2"><a class="reference internal" href="plot/pertpy.plot.guide.html">pertpy.plot.guide</a></li>
<li class="toctree-l2"><a class="reference internal" href="plot/pertpy.plot.ms.violin.html">pertpy.plot.ms.violin</a></li>
<li class="toctree-l2"><a class="reference internal" href="plot/pertpy.plot.ms.perturbscore.html">pertpy.plot.ms.perturbscore</a></li>
<li class="toctree-l2"><a class="reference internal" href="plot/pertpy.plot.ms.heatmap.html">pertpy.plot.ms.heatmap</a></li>
<li class="toctree-l2"><a class="reference internal" href="plot/pertpy.plot.ms.barplot.html">pertpy.plot.ms.barplot</a></li>
<li class="toctree-l2"><a class="reference internal" href="plot/pertpy.plot.ms.lda.html">pertpy.plot.ms.lda</a></li>
<li class="toctree-l2"><a class="reference internal" href="plot/pertpy.plot.milo.nhood_graph.html">pertpy.plot.milo.nhood_graph</a></li>
<li class="toctree-l2"><a class="reference internal" href="plot/pertpy.plot.milo.nhood.html">pertpy.plot.milo.nhood</a></li>
<li class="toctree-l2"><a class="reference internal" href="plot/pertpy.plot.milo.da_beeswarm.html">pertpy.plot.milo.da_beeswarm</a></li>
<li class="toctree-l2"><a class="reference internal" href="plot/pertpy.plot.milo.nhood_counts_by_cond.html">pertpy.plot.milo.nhood_counts_by_cond</a></li>
<li class="toctree-l2"><a class="reference internal" href="plot/pertpy.plot.coda.stacked_barplot.html">pertpy.plot.coda.stacked_barplot</a></li>
<li class="toctree-l2"><a class="reference internal" href="plot/pertpy.plot.coda.effects_barplot.html">pertpy.plot.coda.effects_barplot</a></li>
<li class="toctree-l2"><a class="reference internal" href="plot/pertpy.plot.coda.boxplots.html">pertpy.plot.coda.boxplots</a></li>
<li class="toctree-l2"><a class="reference internal" href="plot/pertpy.plot.coda.rel_abundance_dispersion_plot.html">pertpy.plot.coda.rel_abundance_dispersion_plot</a></li>
<li class="toctree-l2"><a class="reference internal" href="plot/pertpy.plot.coda.draw_tree.html">pertpy.plot.coda.draw_tree</a></li>
<li class="toctree-l2"><a class="reference internal" href="plot/pertpy.plot.coda.draw_effects.html">pertpy.plot.coda.draw_effects</a></li>
<li class="toctree-l2"><a class="reference internal" href="plot/pertpy.plot.coda.effects_umap.html">pertpy.plot.coda.effects_umap</a></li>
<li class="toctree-l2"><a class="reference internal" href="plot/pertpy.plot.ag.dp_scatter.html">pertpy.plot.ag.dp_scatter</a></li>
<li class="toctree-l2"><a class="reference internal" href="plot/pertpy.plot.ag.important_features.html">pertpy.plot.ag.important_features</a></li>
<li class="toctree-l2"><a class="reference internal" href="plot/pertpy.plot.ag.lollipop.html">pertpy.plot.ag.lollipop</a></li>
<li class="toctree-l2"><a class="reference internal" href="plot/pertpy.plot.ag.scatterplot.html">pertpy.plot.ag.scatterplot</a></li>
<li class="toctree-l2"><a class="reference internal" href="plot/pertpy.plot.scg.reg_mean_plot.html">pertpy.plot.scg.reg_mean_plot</a></li>
<li class="toctree-l2"><a class="reference internal" href="plot/pertpy.plot.scg.reg_var_plot.html">pertpy.plot.scg.reg_var_plot</a></li>
<li class="toctree-l2"><a class="reference internal" href="plot/pertpy.plot.scg.binary_classifier.html">pertpy.plot.scg.binary_classifier</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/index.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../contributing.html">Contributing guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../code_of_conduct.html">Contributor Covenant Code of Conduct</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/theislab/pertpy/discussions">Discussions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../authors.html">Credits</a></li>
<li class="toctree-l1"><a class="reference internal" href="../references.html">References</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <section class="tex2jax_ignore mathjax_ignore" id="usage">
<h1>Usage<a class="headerlink" href="#usage" title="Link to this heading">#</a></h1>
<p>Import the pertpy API as follows:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pertpy</span> <span class="k">as</span> <span class="nn">pt</span>
</pre></div>
</div>
<p>You can then access the respective modules like:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">pt</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">cool_fancy_plot</span><span class="p">()</span>
</pre></div>
</div>
<section id="datasets">
<h2>Datasets<a class="headerlink" href="#datasets" title="Link to this heading">#</a></h2>
<div class="table-wrapper autosummary longtable docutils container">
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.adamson_2016_pilot.html#pertpy.data.adamson_2016_pilot" title="pertpy.data.adamson_2016_pilot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.adamson_2016_pilot</span></code></a>()</p></td>
<td><p>6000 chronic myeloid leukemia (K562) cells carrying 8 distinct GBCs.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.adamson_2016_upr_epistasis.html#pertpy.data.adamson_2016_upr_epistasis" title="pertpy.data.adamson_2016_upr_epistasis"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.adamson_2016_upr_epistasis</span></code></a>()</p></td>
<td><p>15000 K562 cells with UPR sensor genes knocked out and treated with thapsigargin.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.adamson_2016_upr_perturb_seq.html#pertpy.data.adamson_2016_upr_perturb_seq" title="pertpy.data.adamson_2016_upr_perturb_seq"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.adamson_2016_upr_perturb_seq</span></code></a>()</p></td>
<td><p>Transcriptomics measurements of 65000 cells that were subject to 91 sgRNAs targeting 82 genes.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.aissa_2021.html#pertpy.data.aissa_2021" title="pertpy.data.aissa_2021"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.aissa_2021</span></code></a>()</p></td>
<td><p>Transcriptomics of 848 P99 cells subject to consecutive erlotinib and 756 control cells.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.bhattacherjee.html#pertpy.data.bhattacherjee" title="pertpy.data.bhattacherjee"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.bhattacherjee</span></code></a>()</p></td>
<td><p>Processed single-cell data PFC adult mice under cocaine self-administration.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.burczynski_crohn.html#pertpy.data.burczynski_crohn" title="pertpy.data.burczynski_crohn"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.burczynski_crohn</span></code></a>()</p></td>
<td><p>Bulk data with conditions ulcerative colitis (UC) and Crohn's disease (CD).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.chang_2021.html#pertpy.data.chang_2021" title="pertpy.data.chang_2021"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.chang_2021</span></code></a>()</p></td>
<td><p>Transcriptomics of 5 different cell lines that were induced with a unique TraCe-seq barcode.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.datlinger_2017.html#pertpy.data.datlinger_2017" title="pertpy.data.datlinger_2017"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.datlinger_2017</span></code></a>()</p></td>
<td><p>Transcriptomics measurements of 5905 Jurkat cells induced with anti-CD3 and anti-CD28 antibodies.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.datlinger_2021.html#pertpy.data.datlinger_2021" title="pertpy.data.datlinger_2021"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.datlinger_2021</span></code></a>()</p></td>
<td><p>Transcriptomics measurements of 151788 nuclei of four cell lines.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.dialogue_example.html#pertpy.data.dialogue_example" title="pertpy.data.dialogue_example"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.dialogue_example</span></code></a>()</p></td>
<td><p>Example dataset used in DIALOGUE vignettes.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.dixit_2016.html#pertpy.data.dixit_2016" title="pertpy.data.dixit_2016"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.dixit_2016</span></code></a>()</p></td>
<td><p>Perturb-seq: scRNA-seq with pooled CRISPR-KO perturbations.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.dixit_2016_raw.html#pertpy.data.dixit_2016_raw" title="pertpy.data.dixit_2016_raw"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.dixit_2016_raw</span></code></a>()</p></td>
<td><p>Perturb-seq: scRNA-seq with pooled CRISPR-KO perturbations.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.frangieh_2021.html#pertpy.data.frangieh_2021" title="pertpy.data.frangieh_2021"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.frangieh_2021</span></code></a>()</p></td>
<td><p>Processed perturb-CITE-seq data with multi-modal RNA and protein single-cell profiling.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.frangieh_2021_protein.html#pertpy.data.frangieh_2021_protein" title="pertpy.data.frangieh_2021_protein"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.frangieh_2021_protein</span></code></a>()</p></td>
<td><p>CITE-seq data of 218000 cells under 750 perturbations (only the surface protein data).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.frangieh_2021_raw.html#pertpy.data.frangieh_2021_raw" title="pertpy.data.frangieh_2021_raw"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.frangieh_2021_raw</span></code></a>()</p></td>
<td><p>Raw Perturb-CITE-seq data with multi-modal RNA and protein single-cell profiling readout.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.frangieh_2021_rna.html#pertpy.data.frangieh_2021_rna" title="pertpy.data.frangieh_2021_rna"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.frangieh_2021_rna</span></code></a>()</p></td>
<td><p>CITE-seq data of 218000 cells under 750 perturbations (only the transcriptomics data).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.gasperini_2019_atscale.html#pertpy.data.gasperini_2019_atscale" title="pertpy.data.gasperini_2019_atscale"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.gasperini_2019_atscale</span></code></a>()</p></td>
<td><p>Transcriptomics of 254974 cells of chronic K562 cells with CRISPRi perturbations.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.gasperini_2019_highmoi.html#pertpy.data.gasperini_2019_highmoi" title="pertpy.data.gasperini_2019_highmoi"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.gasperini_2019_highmoi</span></code></a>()</p></td>
<td><p>K562 perturbed cells with 1119 candidate enhancers (only the high MOI part).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.gasperini_2019_lowmoi.html#pertpy.data.gasperini_2019_lowmoi" title="pertpy.data.gasperini_2019_lowmoi"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.gasperini_2019_lowmoi</span></code></a>()</p></td>
<td><p>K562 perturbed cells with 1119 candidate enhancers (only the low MOI part).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.gehring_2019.html#pertpy.data.gehring_2019" title="pertpy.data.gehring_2019"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.gehring_2019</span></code></a>()</p></td>
<td><p>96-plex perturbation experiment on live mouse neural stem cells.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.haber_2017_regions.html#pertpy.data.haber_2017_regions" title="pertpy.data.haber_2017_regions"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.haber_2017_regions</span></code></a>()</p></td>
<td><p>Raw single-cell, pooled CRISPR screening.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.kang_2018.html#pertpy.data.kang_2018" title="pertpy.data.kang_2018"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.kang_2018</span></code></a>()</p></td>
<td><p>Processed multiplexing droplet-based single cell RNA-sequencing using genetic barcodes.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.mcfarland_2020.html#pertpy.data.mcfarland_2020" title="pertpy.data.mcfarland_2020"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.mcfarland_2020</span></code></a>()</p></td>
<td><p>Response of various cell lines to a range of different drugs and CRISPRi perturbations.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.norman_2019.html#pertpy.data.norman_2019" title="pertpy.data.norman_2019"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.norman_2019</span></code></a>()</p></td>
<td><p>Processed single-cell, pooled CRISPR screening.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.norman_2019_raw.html#pertpy.data.norman_2019_raw" title="pertpy.data.norman_2019_raw"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.norman_2019_raw</span></code></a>()</p></td>
<td><p>Raw single-cell, pooled CRISPR screening.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.papalexi_2021.html#pertpy.data.papalexi_2021" title="pertpy.data.papalexi_2021"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.papalexi_2021</span></code></a>()</p></td>
<td><p>ECCITE-seq dataset of 11 gRNAs generated from stimulated THP-1 cell line.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.replogle_2022_k562_essential.html#pertpy.data.replogle_2022_k562_essential" title="pertpy.data.replogle_2022_k562_essential"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.replogle_2022_k562_essential</span></code></a>()</p></td>
<td><p>K562 cells transduced with CRISPRi (day 7 after transduction).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.replogle_2022_k562_gwps.html#pertpy.data.replogle_2022_k562_gwps" title="pertpy.data.replogle_2022_k562_gwps"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.replogle_2022_k562_gwps</span></code></a>()</p></td>
<td><p>K562 cells transduced with CRISPRi (day 8 after transcduction).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.replogle_2022_rpe1.html#pertpy.data.replogle_2022_rpe1" title="pertpy.data.replogle_2022_rpe1"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.replogle_2022_rpe1</span></code></a>()</p></td>
<td><p>RPE1 cells transduced with CRISPRi (day 7 after transcduction).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.sc_sim_augur.html#pertpy.data.sc_sim_augur" title="pertpy.data.sc_sim_augur"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.sc_sim_augur</span></code></a>()</p></td>
<td><p>Simulated test dataset used in Augur example vignettes.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.schiebinger_2019_16day.html#pertpy.data.schiebinger_2019_16day" title="pertpy.data.schiebinger_2019_16day"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.schiebinger_2019_16day</span></code></a>()</p></td>
<td><p>Transcriptomes of 65781 iPSC cells collected over 10 time points in 2i or serum conditions (16-day time course).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.schiebinger_2019_18day.html#pertpy.data.schiebinger_2019_18day" title="pertpy.data.schiebinger_2019_18day"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.schiebinger_2019_18day</span></code></a>()</p></td>
<td><p>Transcriptomes of 259155 iPSC cells collected over 39 time points in 2i or serum conditions (18-day time course).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.schraivogel_2020_tap_screen_chr8.html#pertpy.data.schraivogel_2020_tap_screen_chr8" title="pertpy.data.schraivogel_2020_tap_screen_chr8"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.schraivogel_2020_tap_screen_chr8</span></code></a>()</p></td>
<td><p>TAP-seq applied to K562 cells (only chromosome 8).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.schraivogel_2020_tap_screen_chr11.html#pertpy.data.schraivogel_2020_tap_screen_chr11" title="pertpy.data.schraivogel_2020_tap_screen_chr11"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.schraivogel_2020_tap_screen_chr11</span></code></a>()</p></td>
<td><p>TAP-seq applied to K562 cells (only chromosome 11).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.sciplex3_raw.html#pertpy.data.sciplex3_raw" title="pertpy.data.sciplex3_raw"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.sciplex3_raw</span></code></a>()</p></td>
<td><p>Raw sciplex3 perturbation dataset curated for perturbation modeling.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.shifrut_2018.html#pertpy.data.shifrut_2018" title="pertpy.data.shifrut_2018"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.shifrut_2018</span></code></a>()</p></td>
<td><p>CD8 T-cells from two donors for two conditions (SLICE and CROP-seq).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.smillie.html#pertpy.data.smillie" title="pertpy.data.smillie"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.smillie</span></code></a>()</p></td>
<td><p>scRNA-seq data of the small intestine of mice under Ulcerative Colitis.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.srivatsan_2020_sciplex2.html#pertpy.data.srivatsan_2020_sciplex2" title="pertpy.data.srivatsan_2020_sciplex2"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.srivatsan_2020_sciplex2</span></code></a>()</p></td>
<td><p>A549 cells exposed to four compounds.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.srivatsan_2020_sciplex3.html#pertpy.data.srivatsan_2020_sciplex3" title="pertpy.data.srivatsan_2020_sciplex3"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.srivatsan_2020_sciplex3</span></code></a>()</p></td>
<td><p>Transcriptomes of 650000 A549, K562, and mCF7 cells exposed to 188 compounds.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.srivatsan_2020_sciplex4.html#pertpy.data.srivatsan_2020_sciplex4" title="pertpy.data.srivatsan_2020_sciplex4"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.srivatsan_2020_sciplex4</span></code></a>()</p></td>
<td><p>A549 and MCF7 cells treated with pracinostat.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.stephenson_2021_subsampled.html#pertpy.data.stephenson_2021_subsampled" title="pertpy.data.stephenson_2021_subsampled"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.stephenson_2021_subsampled</span></code></a>()</p></td>
<td><p>Processed 10X 5' scRNA-seq data from PBMC of COVID-19 patients and healthy donors</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.tian_2019_day7neuron.html#pertpy.data.tian_2019_day7neuron" title="pertpy.data.tian_2019_day7neuron"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.tian_2019_day7neuron</span></code></a>()</p></td>
<td><p>Transcriptomes of 20000 day 7 neurons targeted by 58 gRNAs.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.tian_2019_ipsc.html#pertpy.data.tian_2019_ipsc" title="pertpy.data.tian_2019_ipsc"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.tian_2019_ipsc</span></code></a>()</p></td>
<td><p>Transcriptomics of 20000 iPSCs targeted by 58 sgRNAs.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.tian_2021_crispra.html#pertpy.data.tian_2021_crispra" title="pertpy.data.tian_2021_crispra"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.tian_2021_crispra</span></code></a>()</p></td>
<td><p>CROP-seq of 50000 neurons treated with 374 gRNAs (CRISPRa only).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.tian_2021_crispri.html#pertpy.data.tian_2021_crispri" title="pertpy.data.tian_2021_crispri"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.tian_2021_crispri</span></code></a>()</p></td>
<td><p>CROP-seq of 98000 neurons treated with 374 gRNAs (CRISPRi only).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.weinreb_2020.html#pertpy.data.weinreb_2020" title="pertpy.data.weinreb_2020"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.weinreb_2020</span></code></a>()</p></td>
<td><p>Mouse embryonic stem cells under different cytokines across time.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.xie_2017.html#pertpy.data.xie_2017" title="pertpy.data.xie_2017"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.xie_2017</span></code></a>()</p></td>
<td><p>Single-cell transcriptomics of 51448 cells generated with Mosaic-seq.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.zhao_2021.html#pertpy.data.zhao_2021" title="pertpy.data.zhao_2021"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.zhao_2021</span></code></a>()</p></td>
<td><p>Multiplexed drug perturbation from freshly resected tumors.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="preprocessing">
<h2>Preprocessing<a class="headerlink" href="#preprocessing" title="Link to this heading">#</a></h2>
<section id="guide-assignment">
<h3>Guide Assignment<a class="headerlink" href="#guide-assignment" title="Link to this heading">#</a></h3>
<p>Simple functions for:</p>
<p>Assigning guides based on thresholds. Each cell is assigned to the most expressed gRNA if it has at least the specified number of counts.</p>
<div class="table-wrapper autosummary longtable docutils container">
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="preprocessing/pertpy.preprocessing.GuideAssignment.html#pertpy.preprocessing.GuideAssignment" title="pertpy.preprocessing.GuideAssignment"><code class="xref py py-obj docutils literal notranslate"><span class="pre">preprocessing.GuideAssignment</span></code></a></p></td>
<td><p>Offers simple guide assigment based on count thresholds.</p></td>
</tr>
</tbody>
</table>
</div>
<p>Example implementation:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pertpy</span> <span class="k">as</span> <span class="nn">pt</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>

<span class="n">mdata</span> <span class="o">=</span> <span class="n">pt</span><span class="o">.</span><span class="n">dt</span><span class="o">.</span><span class="n">papalexi_2021</span><span class="p">()</span>
<span class="n">gdo</span> <span class="o">=</span> <span class="n">mdata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="s2">&quot;gdo&quot;</span><span class="p">]</span>
<span class="n">gdo</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;counts&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gdo</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">gdo</span><span class="p">)</span>

<span class="n">ga</span> <span class="o">=</span> <span class="n">pt</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">GuideAssignment</span><span class="p">()</span>
<span class="n">ga</span><span class="o">.</span><span class="n">assign_by_threshold</span><span class="p">(</span><span class="n">gdo</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="n">layer</span><span class="o">=</span><span class="s2">&quot;counts&quot;</span><span class="p">,</span> <span class="n">output_layer</span><span class="o">=</span><span class="s2">&quot;assigned_guides&quot;</span><span class="p">)</span>

<span class="n">pt</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">guide</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">gdo</span><span class="p">,</span> <span class="n">layer</span><span class="o">=</span><span class="s2">&quot;assigned_guides&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<section id="tools">
<h2>Tools<a class="headerlink" href="#tools" title="Link to this heading">#</a></h2>
<section id="pooled-crispr-screens">
<h3>Pooled CRISPR screens<a class="headerlink" href="#pooled-crispr-screens" title="Link to this heading">#</a></h3>
<section id="mixscape">
<h4>Mixscape<a class="headerlink" href="#mixscape" title="Link to this heading">#</a></h4>
<p>A Python implementation of <a class="reference external" href="https://satijalab.org/seurat/articles/mixscape_vignette.html">Mixscape</a>
Papalexi et al. <a class="reference external" href="https://www.nature.com/articles/s41588-021-00778-2">Characterizing the molecular regulation of inhibitory immune checkpoints with multimodal single-cell screens</a>.</p>
<p>Mixscape first tries to remove confounding sources of variation such as cell cycle or replicate effect by embedding the cells into a perturbation space (the perturbation signature).
Next, it determines which targeted cells were affected by the genetic perturbation (=KO) and which targeted cells were not (=NP) with the use of mixture models.
Finally, it visualizes similarities and differences across different perturbations.</p>
<div class="table-wrapper autosummary longtable docutils container">
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="tools/pertpy.tools.Mixscape.html#pertpy.tools.Mixscape" title="pertpy.tools.Mixscape"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tools.Mixscape</span></code></a>()</p></td>
<td><p>Python implementation of Mixscape.</p></td>
</tr>
</tbody>
</table>
</div>
<p>Example implementation:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pertpy</span> <span class="k">as</span> <span class="nn">pt</span>

<span class="n">mdata</span> <span class="o">=</span> <span class="n">pt</span><span class="o">.</span><span class="n">dt</span><span class="o">.</span><span class="n">papalexi_2021</span><span class="p">()</span>
<span class="n">ms</span> <span class="o">=</span> <span class="n">pt</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">Mixscape</span><span class="p">()</span>
<span class="n">ms</span><span class="o">.</span><span class="n">perturbation_signature</span><span class="p">(</span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;rna&quot;</span><span class="p">],</span> <span class="s2">&quot;perturbation&quot;</span><span class="p">,</span> <span class="s2">&quot;NT&quot;</span><span class="p">,</span> <span class="s2">&quot;replicate&quot;</span><span class="p">)</span>
<span class="n">ms</span><span class="o">.</span><span class="n">mixscape</span><span class="p">(</span><span class="n">adata</span><span class="o">=</span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;rna&quot;</span><span class="p">],</span> <span class="n">control</span><span class="o">=</span><span class="s2">&quot;NT&quot;</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="s2">&quot;gene_target&quot;</span><span class="p">,</span> <span class="n">layer</span><span class="o">=</span><span class="s2">&quot;X_pert&quot;</span><span class="p">)</span>
<span class="n">ms</span><span class="o">.</span><span class="n">lda</span><span class="p">(</span><span class="n">adata</span><span class="o">=</span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;rna&quot;</span><span class="p">],</span> <span class="n">labels</span><span class="o">=</span><span class="s2">&quot;gene_target&quot;</span><span class="p">,</span> <span class="n">layer</span><span class="o">=</span><span class="s2">&quot;X_pert&quot;</span><span class="p">)</span>
<span class="n">pt</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">ms</span><span class="o">.</span><span class="n">lda</span><span class="p">(</span><span class="n">adata</span><span class="o">=</span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;rna&quot;</span><span class="p">])</span>
</pre></div>
</div>
<p>See <a class="reference external" href="https://pertpy.readthedocs.io/en/latest/tutorials/notebooks/mixscape.html">mixscape tutorial</a> for a more elaborate tutorial.</p>
</section>
</section>
<section id="compositional-analysis">
<h3>Compositional analysis<a class="headerlink" href="#compositional-analysis" title="Link to this heading">#</a></h3>
<section id="milo">
<h4>Milo<a class="headerlink" href="#milo" title="Link to this heading">#</a></h4>
<p>A Python implementation of Milo for differential abundance testing on KNN graphs, to ease interoperability with scverse pipelines for single-cell analysis.
See <a class="reference external" href="https://www.nature.com/articles/s41587-021-01033-z">Differential abundance testing on single-cell data using k-nearest neighbor graphs</a> for details on the statistical framework.</p>
<div class="table-wrapper autosummary longtable docutils container">
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="tools/pertpy.tools.Milo.html#pertpy.tools.Milo" title="pertpy.tools.Milo"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tools.Milo</span></code></a>()</p></td>
<td><p>Python implementation of Milo.</p></td>
</tr>
</tbody>
</table>
</div>
<p>See <a class="reference external" href="https://pertpy.readthedocs.io/en/latest/tutorials/notebooks/milo.html">milo tutorial</a> for a more elaborate tutorial.</p>
<p>Example implementation:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pertpy</span> <span class="k">as</span> <span class="nn">pt</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>

<span class="n">adata</span> <span class="o">=</span> <span class="n">pt</span><span class="o">.</span><span class="n">dt</span><span class="o">.</span><span class="n">stephenson_2021_subsampled</span><span class="p">()</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;COVID_severity&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;Status_on_day_collection_summary&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[[</span><span class="s2">&quot;patient_id&quot;</span><span class="p">,</span> <span class="s2">&quot;COVID_severity&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;Status&quot;</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;LPS&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

<span class="n">milo</span> <span class="o">=</span> <span class="n">pt</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">Milo</span><span class="p">()</span>
<span class="n">mdata</span> <span class="o">=</span> <span class="n">milo</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;rna&quot;</span><span class="p">],</span> <span class="n">use_rep</span><span class="o">=</span><span class="s2">&quot;X_scVI&quot;</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">150</span><span class="p">,</span> <span class="n">n_pcs</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="n">milo</span><span class="o">.</span><span class="n">make_nhoods</span><span class="p">(</span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;rna&quot;</span><span class="p">],</span> <span class="n">prop</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>
<span class="n">mdata</span> <span class="o">=</span> <span class="n">milo</span><span class="o">.</span><span class="n">count_nhoods</span><span class="p">(</span><span class="n">mdata</span><span class="p">,</span> <span class="n">sample_col</span><span class="o">=</span><span class="s2">&quot;patient_id&quot;</span><span class="p">)</span>
<span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;rna&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;Status&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;rna&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;Status&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">reorder_categories</span><span class="p">([</span><span class="s2">&quot;Healthy&quot;</span><span class="p">,</span> <span class="s2">&quot;Covid&quot;</span><span class="p">])</span>
<span class="p">)</span>
<span class="n">milo</span><span class="o">.</span><span class="n">da_nhoods</span><span class="p">(</span><span class="n">mdata</span><span class="p">,</span> <span class="n">design</span><span class="o">=</span><span class="s2">&quot;~Status&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="sccoda-and-tasccoda">
<h4>scCODA and tascCODA<a class="headerlink" href="#sccoda-and-tasccoda" title="Link to this heading">#</a></h4>
<p>Reimplementation of scCODA for identification of compositional changes in high-throughput sequencing count data and tascCODA for sparse, tree-aggregated modeling of high-throughput sequencing data.
See <a class="reference external" href="https://www.nature.com/articles/s41467-021-27150-6">scCODA is a Bayesian model for compositional single-cell data analysis</a> for statistical methodology and benchmarking performance of scCODA and <a class="reference external" href="https://www.frontiersin.org/articles/10.3389/fgene.2021.766405/full">tascCODA: Bayesian Tree-Aggregated Analysis of Compositional Amplicon and Single-Cell Data</a> for statistical methodology and benchmarking performance of tascCODA.</p>
<div class="table-wrapper autosummary longtable docutils container">
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="tools/pertpy.tools.Sccoda.html#pertpy.tools.Sccoda" title="pertpy.tools.Sccoda"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tools.Sccoda</span></code></a>(*args, **kwargs)</p></td>
<td><p>Statistical model for single-cell differential composition analysis with specification of a reference cell type.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="tools/pertpy.tools.Tasccoda.html#pertpy.tools.Tasccoda" title="pertpy.tools.Tasccoda"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tools.Tasccoda</span></code></a>(*args, **kwargs)</p></td>
<td><p>Statistical model for tree-aggregated differential composition analysis (tascCODA, Ostner et al., 2021).</p></td>
</tr>
</tbody>
</table>
</div>
<p>Example implementation:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pertpy</span> <span class="k">as</span> <span class="nn">pt</span>

<span class="n">haber_cells</span> <span class="o">=</span> <span class="n">pt</span><span class="o">.</span><span class="n">dt</span><span class="o">.</span><span class="n">haber_2017_regions</span><span class="p">()</span>
<span class="n">sccoda_model</span> <span class="o">=</span> <span class="n">pt</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">Sccoda</span><span class="p">()</span>
<span class="n">sccoda_data</span> <span class="o">=</span> <span class="n">sccoda_model</span><span class="o">.</span><span class="n">load</span><span class="p">(</span>
    <span class="n">haber_cells</span><span class="p">,</span>
    <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;cell_level&quot;</span><span class="p">,</span>
    <span class="n">generate_sample_level</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">cell_type_identifier</span><span class="o">=</span><span class="s2">&quot;cell_label&quot;</span><span class="p">,</span>
    <span class="n">sample_identifier</span><span class="o">=</span><span class="s2">&quot;batch&quot;</span><span class="p">,</span>
    <span class="n">covariate_obs</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;condition&quot;</span><span class="p">],</span>
<span class="p">)</span>
<span class="n">sccoda_data</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="s2">&quot;coda_salm&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">sccoda_data</span><span class="p">[</span><span class="s2">&quot;coda&quot;</span><span class="p">][</span>
    <span class="n">sccoda_data</span><span class="p">[</span><span class="s2">&quot;coda&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;condition&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="s2">&quot;Control&quot;</span><span class="p">,</span> <span class="s2">&quot;Salmonella&quot;</span><span class="p">])</span>
<span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

<span class="n">sccoda_data</span> <span class="o">=</span> <span class="n">sccoda_model</span><span class="o">.</span><span class="n">prepare</span><span class="p">(</span>
    <span class="n">sccoda_data</span><span class="p">,</span>
    <span class="n">modality_key</span><span class="o">=</span><span class="s2">&quot;coda_salm&quot;</span><span class="p">,</span>
    <span class="n">formula</span><span class="o">=</span><span class="s2">&quot;condition&quot;</span><span class="p">,</span>
    <span class="n">reference_cell_type</span><span class="o">=</span><span class="s2">&quot;Goblet&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">sccoda_model</span><span class="o">.</span><span class="n">run_nuts</span><span class="p">(</span><span class="n">sccoda_data</span><span class="p">,</span> <span class="n">modality_key</span><span class="o">=</span><span class="s2">&quot;coda_salm&quot;</span><span class="p">)</span>
<span class="n">sccoda_model</span><span class="o">.</span><span class="n">summary</span><span class="p">(</span><span class="n">sccoda_data</span><span class="p">,</span> <span class="n">modality_key</span><span class="o">=</span><span class="s2">&quot;coda_salm&quot;</span><span class="p">)</span>
<span class="n">pt</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">coda</span><span class="o">.</span><span class="n">effects_barplot</span><span class="p">(</span>
    <span class="n">sccoda_data</span><span class="p">,</span> <span class="n">modality_key</span><span class="o">=</span><span class="s2">&quot;coda_salm&quot;</span><span class="p">,</span> <span class="n">parameter</span><span class="o">=</span><span class="s2">&quot;Final Parameter&quot;</span>
<span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<section id="multi-cellular-or-gene-programs">
<h3>Multi-cellular or gene programs<a class="headerlink" href="#multi-cellular-or-gene-programs" title="Link to this heading">#</a></h3>
<section id="dialogue">
<h4>DIALOGUE<a class="headerlink" href="#dialogue" title="Link to this heading">#</a></h4>
<p>A <strong>work in progress (!)</strong> Python implementation of DIALOGUE for the discovery of multicellular programs.
See <a class="reference external" href="https://www.nature.com/articles/s41587-022-01288-0">DIALOGUE maps multicellular programs in tissue from single-cell or spatial transcriptomics data</a> for more details on the methodology.</p>
<div class="table-wrapper autosummary longtable docutils container">
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="tools/pertpy.tools.Dialogue.html#pertpy.tools.Dialogue" title="pertpy.tools.Dialogue"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tools.Dialogue</span></code></a>(sample_id, celltype_key, ...)</p></td>
<td><p>Python implementation of DIALOGUE</p></td>
</tr>
</tbody>
</table>
</div>
<p>See <a class="reference external" href="https://pertpy.readthedocs.io/en/latest/tutorials/notebooks/dialogue.html">dialogue tutorial</a> for a more elaborate tutorial.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pertpy</span> <span class="k">as</span> <span class="nn">pt</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>

<span class="n">adata</span> <span class="o">=</span> <span class="n">pt</span><span class="o">.</span><span class="n">dt</span><span class="o">.</span><span class="n">dialogue_example</span><span class="p">()</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>


<span class="n">dl</span> <span class="o">=</span> <span class="n">pt</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">Dialogue</span><span class="p">(</span>
    <span class="n">sample_id</span><span class="o">=</span><span class="s2">&quot;clinical.status&quot;</span><span class="p">,</span>
    <span class="n">celltype_key</span><span class="o">=</span><span class="s2">&quot;cell.subtypes&quot;</span><span class="p">,</span>
    <span class="n">n_counts_key</span><span class="o">=</span><span class="s2">&quot;nCount_RNA&quot;</span><span class="p">,</span>
    <span class="n">n_mpcs</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">adata</span><span class="p">,</span> <span class="n">mcps</span><span class="p">,</span> <span class="n">ws</span><span class="p">,</span> <span class="n">ct_subs</span> <span class="o">=</span> <span class="n">dl</span><span class="o">.</span><span class="n">calculate_multifactor_PMD</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">normalize</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">all_results</span><span class="p">,</span> <span class="n">new_mcps</span> <span class="o">=</span> <span class="n">dl</span><span class="o">.</span><span class="n">multilevel_modeling</span><span class="p">(</span>
    <span class="n">ct_subs</span><span class="o">=</span><span class="n">ct_subs</span><span class="p">,</span>
    <span class="n">mcp_scores</span><span class="o">=</span><span class="n">mcps</span><span class="p">,</span>
    <span class="n">ws_dict</span><span class="o">=</span><span class="n">ws</span><span class="p">,</span>
    <span class="n">confounder</span><span class="o">=</span><span class="s2">&quot;gender&quot;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<section id="distances-and-permutation-tests">
<h3>Distances and Permutation Tests<a class="headerlink" href="#distances-and-permutation-tests" title="Link to this heading">#</a></h3>
<p>General purpose functions for distances and permutation tests.
Reimplements functions from <a class="reference external" href="http://projects.sanderlab.org/scperturb/">scperturb</a> package.</p>
<div class="table-wrapper autosummary longtable docutils container">
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="tools/pertpy.tools.Distance.html#pertpy.tools.Distance" title="pertpy.tools.Distance"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tools.Distance</span></code></a>([metric, obsm_key])</p></td>
<td><p>Distance class, used to compute distances between groups of cells.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="tools/pertpy.tools.DistanceTest.html#pertpy.tools.DistanceTest" title="pertpy.tools.DistanceTest"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tools.DistanceTest</span></code></a>(metric[, n_perms, ...])</p></td>
<td><p>Run permutation tests using a distance of choice between groups of cells.</p></td>
</tr>
</tbody>
</table>
</div>
<p>See <a class="reference external" href="https://pertpy.readthedocs.io/en/latest/tutorials/notebooks/distances.html">Distance tutorial</a>
and <a class="reference external" href="https://pertpy.readthedocs.io/en/latest/tutorials/notebooks/distance_tests.html">Permutation test tutorial</a> for a more elaborate tutorial.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pertpy</span> <span class="k">as</span> <span class="nn">pt</span>

<span class="n">adata</span> <span class="o">=</span> <span class="n">pt</span><span class="o">.</span><span class="n">dt</span><span class="o">.</span><span class="n">distance_example</span><span class="p">()</span>

<span class="c1"># Pairwise distances</span>
<span class="n">distance</span> <span class="o">=</span> <span class="n">pt</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">Distance</span><span class="p">(</span><span class="n">metric</span><span class="o">=</span><span class="s2">&quot;edistance&quot;</span><span class="p">,</span> <span class="n">obsm_key</span><span class="o">=</span><span class="s2">&quot;X_pca&quot;</span><span class="p">)</span>
<span class="n">pairwise_edistance</span> <span class="o">=</span> <span class="n">distance</span><span class="o">.</span><span class="n">pairwise</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;perturbation&quot;</span><span class="p">)</span>

<span class="c1"># E-test (Permutation test using E-distance)</span>
<span class="n">etest</span> <span class="o">=</span> <span class="n">pt</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">PermutationTest</span><span class="p">(</span>
    <span class="n">metric</span><span class="o">=</span><span class="s2">&quot;edistance&quot;</span><span class="p">,</span> <span class="n">obsm_key</span><span class="o">=</span><span class="s2">&quot;X_pca&quot;</span><span class="p">,</span> <span class="n">correction</span><span class="o">=</span><span class="s2">&quot;holm-sidak&quot;</span>
<span class="p">)</span>
<span class="n">tab</span> <span class="o">=</span> <span class="n">etest</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;perturbation&quot;</span><span class="p">,</span> <span class="n">contrast</span><span class="o">=</span><span class="s2">&quot;control&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="metadata">
<h3>MetaData<a class="headerlink" href="#metadata" title="Link to this heading">#</a></h3>
<p>MetaData provides tooling to fetch and add more metadata to perturbations by querying a couple of databases.
We are currently implementing several sources with more to come.</p>
<p>CellLineMetaData aims to retrieve various types of information related to cell lines, including cell line annotation,
bulk RNA and protein expression data.</p>
<p>Available databases for cell line metadata:</p>
<ul class="simple">
<li><p>The Cancer Dependency Map Project at Broad</p></li>
<li><p>The Cancer Dependency Map Project at Sanger</p></li>
</ul>
<div class="table-wrapper autosummary longtable docutils container">
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="tools/pertpy.tools.CellLineMetaData.html#pertpy.tools.CellLineMetaData" title="pertpy.tools.CellLineMetaData"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tools.CellLineMetaData</span></code></a>()</p></td>
<td><p>Utilities to fetch cell line metadata.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="response-prediction">
<h3>Response prediction<a class="headerlink" href="#response-prediction" title="Link to this heading">#</a></h3>
<section id="augur">
<h4>Augur<a class="headerlink" href="#augur" title="Link to this heading">#</a></h4>
<p>The Python implementation of <a class="reference external" href="https://github.com/neurorestore/Augur">Augur R package</a>
Skinnider, M.A., Squair, J.W., Kathe, C. et al. <a class="reference external" href="https://doi.org/10.1038/s41587-020-0605-1">Cell type prioritization in single-cell data</a>. Nat Biotechnol 39, 3034 (2021).</p>
<p>Augur aims to rank or prioritize cell types according to their response to experimental perturbations given high dimensional single-cell sequencing data.
The basic idea is that in the space of molecular measurements cells reacting heavily to induced perturbations are
more easily separated into perturbed and unperturbed than cell types with little or no response.
This separability is quantified by measuring how well experimental labels (eg. treatment and control) can be predicted within each cell type.
Augur trains a machine learning model predicting experimental labels for each cell type in multiple cross validation runs and
then prioritizes cell type response according to metric scores measuring the accuracy of the model.
For categorical data the area under the curve is the default metric and for numerical data the concordance correlation coefficient
is used as a proxy for how accurate the model is which in turn approximates perturbation response.</p>
<p>Example implementation:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pertpy</span> <span class="k">as</span> <span class="nn">pt</span>

<span class="n">adata</span> <span class="o">=</span> <span class="n">pt</span><span class="o">.</span><span class="n">dt</span><span class="o">.</span><span class="n">sc_sim_augur</span><span class="p">()</span>
<span class="n">ag</span> <span class="o">=</span> <span class="n">pt</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">Augur</span><span class="p">(</span><span class="n">estimator</span><span class="o">=</span><span class="s2">&quot;random_forest_classifier&quot;</span><span class="p">)</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">ag</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">adata</span><span class="p">,</span> <span class="n">results</span> <span class="o">=</span> <span class="n">ag</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>

<span class="c1"># metrics for each cell type</span>
<span class="n">results</span><span class="p">[</span><span class="s2">&quot;summary_metrics&quot;</span><span class="p">]</span>
</pre></div>
</div>
<p>See <a class="reference external" href="https://pertpy.readthedocs.io/en/latest/tutorials/notebooks/augur.html">augur tutorial</a> for a more elaborate tutorial.</p>
<div class="table-wrapper autosummary longtable docutils container">
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="tools/pertpy.tools.Augur.html#pertpy.tools.Augur" title="pertpy.tools.Augur"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tools.Augur</span></code></a></p></td>
<td><p>Python implementation of Augur.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="scgen">
<h4>scGen<a class="headerlink" href="#scgen" title="Link to this heading">#</a></h4>
<p>Reimplementation of scGen for perturbation response prediction of scRNA-seq data in Jax.
See <a class="reference external" href="https://www.nature.com/articles/s41592-019-0494-8">scGen predicts single-cell perturbation responses</a> for more details.</p>
<div class="table-wrapper autosummary longtable docutils container">
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="tools/pertpy.tools.SCGEN.html#pertpy.tools.SCGEN" title="pertpy.tools.SCGEN"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tools.SCGEN</span></code></a>(adata[, n_hidden, n_latent, ...])</p></td>
<td><p>Jax Implementation of scGen model for batch removal and perturbation prediction.</p></td>
</tr>
</tbody>
</table>
</div>
<p>Example implementation:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pertpy</span> <span class="k">as</span> <span class="nn">pt</span>

<span class="n">train</span> <span class="o">=</span> <span class="n">pt</span><span class="o">.</span><span class="n">dt</span><span class="o">.</span><span class="n">kang_2018</span><span class="p">()</span>

<span class="n">train_new</span> <span class="o">=</span> <span class="n">train</span><span class="p">[</span>
    <span class="o">~</span><span class="p">((</span><span class="n">train</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;CD4T&quot;</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">train</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;condition&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;stimulated&quot;</span><span class="p">))</span>
<span class="p">]</span>
<span class="n">train_new</span> <span class="o">=</span> <span class="n">train_new</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

<span class="n">pt</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">SCGEN</span><span class="o">.</span><span class="n">setup_anndata</span><span class="p">(</span><span class="n">train_new</span><span class="p">,</span> <span class="n">batch_key</span><span class="o">=</span><span class="s2">&quot;condition&quot;</span><span class="p">,</span> <span class="n">labels_key</span><span class="o">=</span><span class="s2">&quot;cell_type&quot;</span><span class="p">)</span>
<span class="n">model</span> <span class="o">=</span> <span class="n">pt</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">SCGEN</span><span class="p">(</span><span class="n">train_new</span><span class="p">)</span>
<span class="n">model</span><span class="o">.</span><span class="n">train</span><span class="p">(</span><span class="n">max_epochs</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">batch_size</span><span class="o">=</span><span class="mi">32</span><span class="p">)</span>

<span class="n">pred</span><span class="p">,</span> <span class="n">delta</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span>
    <span class="n">ctrl_key</span><span class="o">=</span><span class="s2">&quot;control&quot;</span><span class="p">,</span> <span class="n">stim_key</span><span class="o">=</span><span class="s2">&quot;stimulated&quot;</span><span class="p">,</span> <span class="n">celltype_to_predict</span><span class="o">=</span><span class="s2">&quot;CD4T&quot;</span>
<span class="p">)</span>
<span class="n">pred</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;condition&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;pred&quot;</span>
</pre></div>
</div>
<p>See <a class="reference external" href="https://pertpy.readthedocs.io/en/latest/tutorials/notebooks/scgen_perturbation_prediction.html">scgen tutorial</a> for a more elaborate tutorial.</p>
</section>
</section>
<section id="perturbation-space">
<h3>Perturbation space<a class="headerlink" href="#perturbation-space" title="Link to this heading">#</a></h3>
<p>Various modules for calculating and evaluating perturbation spaces.</p>
<div class="table-wrapper autosummary longtable docutils container">
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="tools/pertpy.tools.DiscriminatorClassifierSpace.html#pertpy.tools.DiscriminatorClassifierSpace" title="pertpy.tools.DiscriminatorClassifierSpace"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tools.DiscriminatorClassifierSpace</span></code></a>()</p></td>
<td><p>Leveraging discriminator classifier.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="tools/pertpy.tools.CentroidSpace.html#pertpy.tools.CentroidSpace" title="pertpy.tools.CentroidSpace"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tools.CentroidSpace</span></code></a>()</p></td>
<td><p>Computes the centroids per perturbation of a pre-computed embedding.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="tools/pertpy.tools.DBSCANSpace.html#pertpy.tools.DBSCANSpace" title="pertpy.tools.DBSCANSpace"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tools.DBSCANSpace</span></code></a>()</p></td>
<td><p>Cluster the given data using DBSCAN</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="tools/pertpy.tools.KMeansSpace.html#pertpy.tools.KMeansSpace" title="pertpy.tools.KMeansSpace"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tools.KMeansSpace</span></code></a>()</p></td>
<td><p>Computes K-Means clustering of the expression values.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="tools/pertpy.tools.PseudobulkSpace.html#pertpy.tools.PseudobulkSpace" title="pertpy.tools.PseudobulkSpace"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tools.PseudobulkSpace</span></code></a>()</p></td>
<td><p>Determines pseudobulks using decoupler.</p></td>
</tr>
</tbody>
</table>
</div>
<p>Example implementation:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pertpy</span> <span class="k">as</span> <span class="nn">pt</span>

<span class="c1"># TODO will be added soon.</span>
</pre></div>
</div>
<p>See <a class="reference external" href="https://pertpy.readthedocs.io/en/latest/tutorials/notebooks/perturbation_space.html">perturbation space tutorial</a> for a more elaborate tutorial.</p>
</section>
</section>
<section id="plots">
<h2>Plots<a class="headerlink" href="#plots" title="Link to this heading">#</a></h2>
<section id="id1">
<h3>Preprocessing<a class="headerlink" href="#id1" title="Link to this heading">#</a></h3>
<div class="table-wrapper autosummary longtable docutils container">
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="plot/pertpy.plot.guide.html#pertpy.plot.guide" title="pertpy.plot.guide"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot.guide</span></code></a></p></td>
<td><p>alias of <code class="xref py py-class docutils literal notranslate"><span class="pre">GuideRnaPlot</span></code></p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="id2">
<h3>Pooled CRISPR screens<a class="headerlink" href="#id2" title="Link to this heading">#</a></h3>
<section id="id3">
<h4>Mixscape<a class="headerlink" href="#id3" title="Link to this heading">#</a></h4>
<div class="table-wrapper autosummary longtable docutils container">
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="plot/pertpy.plot.ms.violin.html#pertpy.plot.ms.violin" title="pertpy.plot.ms.violin"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot.ms.violin</span></code></a>(adata, target_gene_idents[, ...])</p></td>
<td><p>Violin plot using mixscape results.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="plot/pertpy.plot.ms.perturbscore.html#pertpy.plot.ms.perturbscore" title="pertpy.plot.ms.perturbscore"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot.ms.perturbscore</span></code></a>(adata, labels, target_gene)</p></td>
<td><p>Density plots to visualize perturbation scores calculated by the <cite>pt.tl.mixscape</cite> function.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="plot/pertpy.plot.ms.heatmap.html#pertpy.plot.ms.heatmap" title="pertpy.plot.ms.heatmap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot.ms.heatmap</span></code></a>(adata, labels, target_gene, ...)</p></td>
<td><p>Heatmap plot using mixscape results.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="plot/pertpy.plot.ms.barplot.html#pertpy.plot.ms.barplot" title="pertpy.plot.ms.barplot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot.ms.barplot</span></code></a>(adata, guide_rna_column[, ...])</p></td>
<td><p>Barplot to visualize perturbation scores calculated from RunMixscape function.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="plot/pertpy.plot.ms.lda.html#pertpy.plot.ms.lda" title="pertpy.plot.ms.lda"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot.ms.lda</span></code></a>(adata, control[, ...])</p></td>
<td><p>Visualizing perturbation responses with Linear Discriminant Analysis.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
</section>
<section id="id4">
<h3>Compositional analysis<a class="headerlink" href="#id4" title="Link to this heading">#</a></h3>
<section id="id5">
<h4>Milo<a class="headerlink" href="#id5" title="Link to this heading">#</a></h4>
<div class="table-wrapper autosummary longtable docutils container">
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="plot/pertpy.plot.milo.nhood_graph.html#pertpy.plot.milo.nhood_graph" title="pertpy.plot.milo.nhood_graph"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot.milo.nhood_graph</span></code></a>(mdata[, alpha, ...])</p></td>
<td><p>Visualize DA results on abstracted graph (wrapper around sc.pl.embedding)</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="plot/pertpy.plot.milo.nhood.html#pertpy.plot.milo.nhood" title="pertpy.plot.milo.nhood"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot.milo.nhood</span></code></a>(mdata, ix[, feature_key, ...])</p></td>
<td><p>Visualize cells in a neighbourhood.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="plot/pertpy.plot.milo.da_beeswarm.html#pertpy.plot.milo.da_beeswarm" title="pertpy.plot.milo.da_beeswarm"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot.milo.da_beeswarm</span></code></a>(mdata[, feature_key, ...])</p></td>
<td><p>Plot beeswarm plot of logFC against nhood labels</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="plot/pertpy.plot.milo.nhood_counts_by_cond.html#pertpy.plot.milo.nhood_counts_by_cond" title="pertpy.plot.milo.nhood_counts_by_cond"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot.milo.nhood_counts_by_cond</span></code></a>(mdata, test_var)</p></td>
<td><p>Plot boxplot of cell numbers vs condition of interest</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="id6">
<h4>scCODA and tascCODA<a class="headerlink" href="#id6" title="Link to this heading">#</a></h4>
<div class="table-wrapper autosummary longtable docutils container">
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="plot/pertpy.plot.coda.stacked_barplot.html#pertpy.plot.coda.stacked_barplot" title="pertpy.plot.coda.stacked_barplot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot.coda.stacked_barplot</span></code></a>(data, feature_name)</p></td>
<td><p>Plots a stacked barplot for all levels of a covariate or all samples (if feature_name==&quot;samples&quot;).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="plot/pertpy.plot.coda.effects_barplot.html#pertpy.plot.coda.effects_barplot" title="pertpy.plot.coda.effects_barplot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot.coda.effects_barplot</span></code></a>(data[, ...])</p></td>
<td><p>Barplot visualization for effects.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="plot/pertpy.plot.coda.boxplots.html#pertpy.plot.coda.boxplots" title="pertpy.plot.coda.boxplots"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot.coda.boxplots</span></code></a>(data, feature_name[, ...])</p></td>
<td><p>Grouped boxplot visualization. The cell counts for each cell type are shown as a group of boxplots,</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="plot/pertpy.plot.coda.rel_abundance_dispersion_plot.html#pertpy.plot.coda.rel_abundance_dispersion_plot" title="pertpy.plot.coda.rel_abundance_dispersion_plot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot.coda.rel_abundance_dispersion_plot</span></code></a>(data)</p></td>
<td><p>Plots total variance of relative abundance versus minimum relative abundance of all cell types for determination of a reference cell type.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="plot/pertpy.plot.coda.draw_tree.html#pertpy.plot.coda.draw_tree" title="pertpy.plot.coda.draw_tree"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot.coda.draw_tree</span></code></a>(data[, modality_key, ...])</p></td>
<td><p>Plot a tree using input ete3 tree object.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="plot/pertpy.plot.coda.draw_effects.html#pertpy.plot.coda.draw_effects" title="pertpy.plot.coda.draw_effects"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot.coda.draw_effects</span></code></a>(data, covariate[, ...])</p></td>
<td><p>Plot a tree with colored circles on the nodes indicating significant effects with bar plots which indicate leave-level significant effects.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="plot/pertpy.plot.coda.effects_umap.html#pertpy.plot.coda.effects_umap" title="pertpy.plot.coda.effects_umap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot.coda.effects_umap</span></code></a>(data, effect_name, ...)</p></td>
<td><p>Plot a UMAP visualization colored by effect strength.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
</section>
<section id="id7">
<h3>Response prediction<a class="headerlink" href="#id7" title="Link to this heading">#</a></h3>
<section id="id8">
<h4>Augur<a class="headerlink" href="#id8" title="Link to this heading">#</a></h4>
<div class="table-wrapper autosummary longtable docutils container">
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="plot/pertpy.plot.ag.dp_scatter.html#pertpy.plot.ag.dp_scatter" title="pertpy.plot.ag.dp_scatter"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot.ag.dp_scatter</span></code></a>(results[, top_n, ax, ...])</p></td>
<td><p>Plot result of differential prioritization.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="plot/pertpy.plot.ag.important_features.html#pertpy.plot.ag.important_features" title="pertpy.plot.ag.important_features"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot.ag.important_features</span></code></a>(data[, key, ...])</p></td>
<td><p>Plot a lollipop plot of the n features with largest feature importances.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="plot/pertpy.plot.ag.lollipop.html#pertpy.plot.ag.lollipop" title="pertpy.plot.ag.lollipop"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot.ag.lollipop</span></code></a>(data[, key, ax, return_figure])</p></td>
<td><p>Plot a lollipop plot of the mean augur values.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="plot/pertpy.plot.ag.scatterplot.html#pertpy.plot.ag.scatterplot" title="pertpy.plot.ag.scatterplot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot.ag.scatterplot</span></code></a>(results1, results2[, ...])</p></td>
<td><p>Create scatterplot with two augur results.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="id9">
<h4>scGen<a class="headerlink" href="#id9" title="Link to this heading">#</a></h4>
<div class="table-wrapper autosummary longtable docutils container">
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="plot/pertpy.plot.scg.reg_mean_plot.html#pertpy.plot.scg.reg_mean_plot" title="pertpy.plot.scg.reg_mean_plot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot.scg.reg_mean_plot</span></code></a>(adata, condition_key, ...)</p></td>
<td><p>Plots mean matching figure for a set of specific genes.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="plot/pertpy.plot.scg.reg_var_plot.html#pertpy.plot.scg.reg_var_plot" title="pertpy.plot.scg.reg_var_plot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot.scg.reg_var_plot</span></code></a>(adata, condition_key, ...)</p></td>
<td><p>Plots variance matching figure for a set of specific genes.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="plot/pertpy.plot.scg.binary_classifier.html#pertpy.plot.scg.binary_classifier" title="pertpy.plot.scg.binary_classifier"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot.scg.binary_classifier</span></code></a>(scgen, adata, ...)</p></td>
<td><p>Latent space classifier.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
</section>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="data/pertpy.data.adamson_2016_pilot.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">pertpy.data.adamson_2016_pilot</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="../installation.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Installation</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2021, Lukas Heumos, Theislab
            </div>
            Made with 
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">Usage</a><ul>
<li><a class="reference internal" href="#datasets">Datasets</a></li>
<li><a class="reference internal" href="#preprocessing">Preprocessing</a><ul>
<li><a class="reference internal" href="#guide-assignment">Guide Assignment</a></li>
</ul>
</li>
<li><a class="reference internal" href="#tools">Tools</a><ul>
<li><a class="reference internal" href="#pooled-crispr-screens">Pooled CRISPR screens</a><ul>
<li><a class="reference internal" href="#mixscape">Mixscape</a></li>
</ul>
</li>
<li><a class="reference internal" href="#compositional-analysis">Compositional analysis</a><ul>
<li><a class="reference internal" href="#milo">Milo</a></li>
<li><a class="reference internal" href="#sccoda-and-tasccoda">scCODA and tascCODA</a></li>
</ul>
</li>
<li><a class="reference internal" href="#multi-cellular-or-gene-programs">Multi-cellular or gene programs</a><ul>
<li><a class="reference internal" href="#dialogue">DIALOGUE</a></li>
</ul>
</li>
<li><a class="reference internal" href="#distances-and-permutation-tests">Distances and Permutation Tests</a></li>
<li><a class="reference internal" href="#metadata">MetaData</a></li>
<li><a class="reference internal" href="#response-prediction">Response prediction</a><ul>
<li><a class="reference internal" href="#augur">Augur</a></li>
<li><a class="reference internal" href="#scgen">scGen</a></li>
</ul>
</li>
<li><a class="reference internal" href="#perturbation-space">Perturbation space</a></li>
</ul>
</li>
<li><a class="reference internal" href="#plots">Plots</a><ul>
<li><a class="reference internal" href="#id1">Preprocessing</a></li>
<li><a class="reference internal" href="#id2">Pooled CRISPR screens</a><ul>
<li><a class="reference internal" href="#id3">Mixscape</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id4">Compositional analysis</a><ul>
<li><a class="reference internal" href="#id5">Milo</a></li>
<li><a class="reference internal" href="#id6">scCODA and tascCODA</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id7">Response prediction</a><ul>
<li><a class="reference internal" href="#id8">Augur</a></li>
<li><a class="reference internal" href="#id9">scGen</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script src="../_static/documentation_options.js?v=8fa8b3e9"></script>
    <script src="../_static/doctools.js?v=888ff710"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/scripts/furo.js?v=32e29ea5"></script>
    <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../_static/copybutton.js?v=f281be69"></script>
    <script src="../_static/design-tabs.js?v=36754332"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    </body>
</html>