<!doctype html>
<html class="no-js" lang="en">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />
<meta property="og:title" content="Usage" />
<meta property="og:type" content="website" />
<meta property="og:url" content="https://pertpy.readthedocs.io/en/latest/usage/usage.html" />
<meta property="og:site_name" content="pertpy" />
<meta property="og:description" content="Import the pertpy API as follows: You can then access the respective modules like: Data: Tools: Augurpy: The Python implementation of Augur R package Skinnider, M.A., Squair, J.W., Kathe, C. et al...." />
<meta property="og:image" content="https://pertpy.readthedocs.io/en/latest//_static/logo.png" />
<meta property="og:image:alt" content="pertpy" />
<meta name="description" content="Import the pertpy API as follows: You can then access the respective modules like: Data: Tools: Augurpy: The Python implementation of Augur R package Skinnider, M.A., Squair, J.W., Kathe, C. et al...." />
<link rel="index" title="Index" href="../genindex.html" /><link rel="search" title="Search" href="../search.html" /><link rel="next" title="pertpy.data.adamson_2016_pilot" href="data/pertpy.data.adamson_2016_pilot.html" /><link rel="prev" title="Installation" href="../installation.html" />

    <!-- Generated with Sphinx 5.3.0 and Furo 2022.12.07 -->
        <title>Usage - pertpy</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo.css?digest=91d0f0d1c444bdcb17a68e833c7a53903343c195" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sg_gallery.css" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo-extensions.css?digest=30d1aed668e5c3a91c3e3bf6a60b675221979f0e" />
    <link rel="stylesheet" type="text/css" href="../_static/css/override.css" />
    <link rel="stylesheet" type="text/css" href="../_static/css/sphinx_gallery.css" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  --color-brand-primary: #003262;
  --color-brand-content: #003262per;
  --admonition-font-size: var(--font-size-normal);
  --admonition-title-font-size: var(--font-size-normal);
  --code-font-size: var(--font-size--small);
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../index.html"><div class="brand">pertpy</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../index.html">
  
  
  
</a><form class="sidebar-search-container" method="get" action="../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1 current has-children current-page"><a class="current reference internal" href="#">Usage</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.adamson_2016_pilot.html">pertpy.data.adamson_2016_pilot</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.adamson_2016_upr_epistasis.html">pertpy.data.adamson_2016_upr_epistasis</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.adamson_2016_upr_perturb_seq.html">pertpy.data.adamson_2016_upr_perturb_seq</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.aissa_2021.html">pertpy.data.aissa_2021</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.bhattacherjee.html">pertpy.data.bhattacherjee</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.burczynski_crohn.html">pertpy.data.burczynski_crohn</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.chang_2021.html">pertpy.data.chang_2021</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.datlinger_2017.html">pertpy.data.datlinger_2017</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.datlinger_2021.html">pertpy.data.datlinger_2021</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.dialogue_example.html">pertpy.data.dialogue_example</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.dixit_2016.html">pertpy.data.dixit_2016</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.dixit_2016_raw.html">pertpy.data.dixit_2016_raw</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.dixit_2016_scperturb.html">pertpy.data.dixit_2016_scperturb</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.frangieh_2021.html">pertpy.data.frangieh_2021</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.frangieh_2021_protein.html">pertpy.data.frangieh_2021_protein</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.frangieh_2021_raw.html">pertpy.data.frangieh_2021_raw</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.frangieh_2021_rna.html">pertpy.data.frangieh_2021_rna</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.gasperini_2019_atscale.html">pertpy.data.gasperini_2019_atscale</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.gasperini_2019_highmoi.html">pertpy.data.gasperini_2019_highmoi</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.gasperini_2019_lowmoi.html">pertpy.data.gasperini_2019_lowmoi</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.gehring_2019.html">pertpy.data.gehring_2019</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.haber_2017_regions.html">pertpy.data.haber_2017_regions</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.kang_2018.html">pertpy.data.kang_2018</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.mcfarland_2020.html">pertpy.data.mcfarland_2020</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.norman_2019.html">pertpy.data.norman_2019</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.norman_2019_filtered.html">pertpy.data.norman_2019_filtered</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.norman_2019_raw.html">pertpy.data.norman_2019_raw</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.papalexi_2021.html">pertpy.data.papalexi_2021</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.papalexi_2021_eccite_arrayed_protein.html">pertpy.data.papalexi_2021_eccite_arrayed_protein</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.papalexi_2021_eccite_arrayed_rna.html">pertpy.data.papalexi_2021_eccite_arrayed_rna</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.papalexi_2021_eccite_protein.html">pertpy.data.papalexi_2021_eccite_protein</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.papalexi_2021_eccite_rna.html">pertpy.data.papalexi_2021_eccite_rna</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.replogle_2022_k562_essential.html">pertpy.data.replogle_2022_k562_essential</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.replogle_2022_k562_gwps.html">pertpy.data.replogle_2022_k562_gwps</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.replogle_2022_rpe1.html">pertpy.data.replogle_2022_rpe1</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.sc_sim_augur.html">pertpy.data.sc_sim_augur</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.schiebinger_2019_16day.html">pertpy.data.schiebinger_2019_16day</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.schiebinger_2019_18day.html">pertpy.data.schiebinger_2019_18day</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.schraivogel_2020_tap_screen_chr8.html">pertpy.data.schraivogel_2020_tap_screen_chr8</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.schraivogel_2020_tap_screen_chr11.html">pertpy.data.schraivogel_2020_tap_screen_chr11</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.sciplex3_raw.html">pertpy.data.sciplex3_raw</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.shifrut_2018.html">pertpy.data.shifrut_2018</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.smillie.html">pertpy.data.smillie</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.srivatsan_2020_sciplex2.html">pertpy.data.srivatsan_2020_sciplex2</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.srivatsan_2020_sciplex3.html">pertpy.data.srivatsan_2020_sciplex3</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.srivatsan_2020_sciplex4.html">pertpy.data.srivatsan_2020_sciplex4</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.stephenson_2021_subsampled.html">pertpy.data.stephenson_2021_subsampled</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.tian_2019_day7neuron.html">pertpy.data.tian_2019_day7neuron</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.tian_2019_ipsc.html">pertpy.data.tian_2019_ipsc</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.tian_2021_crispra.html">pertpy.data.tian_2021_crispra</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.tian_2021_crispri.html">pertpy.data.tian_2021_crispri</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.weinreb_2020.html">pertpy.data.weinreb_2020</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.xie_2017.html">pertpy.data.xie_2017</a></li>
<li class="toctree-l2"><a class="reference internal" href="data/pertpy.data.zhao_2021.html">pertpy.data.zhao_2021</a></li>
<li class="toctree-l2"><a class="reference internal" href="tools/pertpy.tools.Augurpy.html">pertpy.tools.Augurpy</a></li>
<li class="toctree-l2"><a class="reference internal" href="tools/pertpy.tools.Mixscape.html">pertpy.tools.Mixscape</a></li>
<li class="toctree-l2"><a class="reference internal" href="tools/pertpy.tools.Milopy.html">pertpy.tools.Milopy</a></li>
<li class="toctree-l2"><a class="reference internal" href="tools/pertpy.tools.Sccoda.html">pertpy.tools.Sccoda</a></li>
<li class="toctree-l2"><a class="reference internal" href="tools/pertpy.tools.Tasccoda.html">pertpy.tools.Tasccoda</a></li>
<li class="toctree-l2"><a class="reference internal" href="tools/pertpy.tools.kernel_pca.html">pertpy.tools.kernel_pca</a></li>
<li class="toctree-l2"><a class="reference internal" href="plot/pertpy.plot.ag.dp_scatter.html">pertpy.plot.ag.dp_scatter</a></li>
<li class="toctree-l2"><a class="reference internal" href="plot/pertpy.plot.ag.important_features.html">pertpy.plot.ag.important_features</a></li>
<li class="toctree-l2"><a class="reference internal" href="plot/pertpy.plot.ag.lollipop.html">pertpy.plot.ag.lollipop</a></li>
<li class="toctree-l2"><a class="reference internal" href="plot/pertpy.plot.ag.scatterplot.html">pertpy.plot.ag.scatterplot</a></li>
<li class="toctree-l2"><a class="reference internal" href="plot/pertpy.plot.ms.violin.html">pertpy.plot.ms.violin</a></li>
<li class="toctree-l2"><a class="reference internal" href="plot/pertpy.plot.ms.perturbscore.html">pertpy.plot.ms.perturbscore</a></li>
<li class="toctree-l2"><a class="reference internal" href="plot/pertpy.plot.ms.heatmap.html">pertpy.plot.ms.heatmap</a></li>
<li class="toctree-l2"><a class="reference internal" href="plot/pertpy.plot.ms.barplot.html">pertpy.plot.ms.barplot</a></li>
<li class="toctree-l2"><a class="reference internal" href="plot/pertpy.plot.ms.lda.html">pertpy.plot.ms.lda</a></li>
<li class="toctree-l2"><a class="reference internal" href="plot/pertpy.plot.milo.nhood_graph.html">pertpy.plot.milo.nhood_graph</a></li>
<li class="toctree-l2"><a class="reference internal" href="plot/pertpy.plot.milo.nhood.html">pertpy.plot.milo.nhood</a></li>
<li class="toctree-l2"><a class="reference internal" href="plot/pertpy.plot.milo.da_beeswarm.html">pertpy.plot.milo.da_beeswarm</a></li>
<li class="toctree-l2"><a class="reference internal" href="plot/pertpy.plot.milo.nhood_counts_by_cond.html">pertpy.plot.milo.nhood_counts_by_cond</a></li>
<li class="toctree-l2"><a class="reference internal" href="plot/pertpy.plot.coda.stacked_barplot.html">pertpy.plot.coda.stacked_barplot</a></li>
<li class="toctree-l2"><a class="reference internal" href="plot/pertpy.plot.coda.effects_barplot.html">pertpy.plot.coda.effects_barplot</a></li>
<li class="toctree-l2"><a class="reference internal" href="plot/pertpy.plot.coda.boxplots.html">pertpy.plot.coda.boxplots</a></li>
<li class="toctree-l2"><a class="reference internal" href="plot/pertpy.plot.coda.rel_abundance_dispersion_plot.html">pertpy.plot.coda.rel_abundance_dispersion_plot</a></li>
<li class="toctree-l2"><a class="reference internal" href="plot/pertpy.plot.coda.draw_tree.html">pertpy.plot.coda.draw_tree</a></li>
<li class="toctree-l2"><a class="reference internal" href="plot/pertpy.plot.coda.draw_effects.html">pertpy.plot.coda.draw_effects</a></li>
<li class="toctree-l2"><a class="reference internal" href="plot/pertpy.plot.coda.effects_umap.html">pertpy.plot.coda.effects_umap</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/index.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../contributing.html">Contributor Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../code_of_conduct.html">Contributor Covenant Code of Conduct</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/theislab/pertpy/discussions">Discussions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../authors.html">Credits</a></li>
<li class="toctree-l1"><a class="reference internal" href="../references.html">References</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <section class="tex2jax_ignore mathjax_ignore" id="usage">
<h1>Usage<a class="headerlink" href="#usage" title="Permalink to this heading">#</a></h1>
<p>Import the pertpy API as follows:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pertpy</span> <span class="k">as</span> <span class="nn">pt</span>
</pre></div>
</div>
<p>You can then access the respective modules like:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">pt</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">cool_fancy_plot</span><span class="p">()</span>
</pre></div>
</div>
<section id="data">
<h2>Data<a class="headerlink" href="#data" title="Permalink to this heading">#</a></h2>
<div class="table-wrapper autosummary longtable docutils container">
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.adamson_2016_pilot.html#pertpy.data.adamson_2016_pilot" title="pertpy.data.adamson_2016_pilot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.adamson_2016_pilot</span></code></a>()</p></td>
<td><p>scPerturb Single-Cell Perturbation Data.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.adamson_2016_upr_epistasis.html#pertpy.data.adamson_2016_upr_epistasis" title="pertpy.data.adamson_2016_upr_epistasis"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.adamson_2016_upr_epistasis</span></code></a>()</p></td>
<td><p>scPerturb Single-Cell Perturbation Data.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.adamson_2016_upr_perturb_seq.html#pertpy.data.adamson_2016_upr_perturb_seq" title="pertpy.data.adamson_2016_upr_perturb_seq"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.adamson_2016_upr_perturb_seq</span></code></a>()</p></td>
<td><p>scPerturb Single-Cell Perturbation Data.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.aissa_2021.html#pertpy.data.aissa_2021" title="pertpy.data.aissa_2021"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.aissa_2021</span></code></a>()</p></td>
<td><p>scPerturb Single-Cell Perturbation Data.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.bhattacherjee.html#pertpy.data.bhattacherjee" title="pertpy.data.bhattacherjee"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.bhattacherjee</span></code></a>()</p></td>
<td><p>Processed single-cell data PFC adult mice under cocaine self-administration.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.burczynski_crohn.html#pertpy.data.burczynski_crohn" title="pertpy.data.burczynski_crohn"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.burczynski_crohn</span></code></a>()</p></td>
<td><p>Bulk data with conditions ulcerative colitis (UC) and Crohn's disease (CD).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.chang_2021.html#pertpy.data.chang_2021" title="pertpy.data.chang_2021"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.chang_2021</span></code></a>()</p></td>
<td><p>scPerturb Single-Cell Perturbation Data.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.datlinger_2017.html#pertpy.data.datlinger_2017" title="pertpy.data.datlinger_2017"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.datlinger_2017</span></code></a>()</p></td>
<td><p>scPerturb Single-Cell Perturbation Data.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.datlinger_2021.html#pertpy.data.datlinger_2021" title="pertpy.data.datlinger_2021"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.datlinger_2021</span></code></a>()</p></td>
<td><p>scPerturb Single-Cell Perturbation Data.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.dialogue_example.html#pertpy.data.dialogue_example" title="pertpy.data.dialogue_example"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.dialogue_example</span></code></a>()</p></td>
<td><p>Example dataset used to feature DIALOGUE.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.dixit_2016.html#pertpy.data.dixit_2016" title="pertpy.data.dixit_2016"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.dixit_2016</span></code></a>()</p></td>
<td><p>Perturb-seq: scRNA-seq with pooled CRISPR-KO perturbations.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.dixit_2016_raw.html#pertpy.data.dixit_2016_raw" title="pertpy.data.dixit_2016_raw"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.dixit_2016_raw</span></code></a>()</p></td>
<td><p>Perturb-seq: scRNA-seq with pooled CRISPR-KO perturbations.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.dixit_2016_scperturb.html#pertpy.data.dixit_2016_scperturb" title="pertpy.data.dixit_2016_scperturb"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.dixit_2016_scperturb</span></code></a>()</p></td>
<td><p>scPerturb Single-Cell Perturbation Data.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.frangieh_2021.html#pertpy.data.frangieh_2021" title="pertpy.data.frangieh_2021"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.frangieh_2021</span></code></a>()</p></td>
<td><p>Processed perturb-CITE-seq data with multi-modal RNA and protein single-cell profiling.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.frangieh_2021_protein.html#pertpy.data.frangieh_2021_protein" title="pertpy.data.frangieh_2021_protein"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.frangieh_2021_protein</span></code></a>()</p></td>
<td><p>scPerturb Single-Cell Perturbation Data.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.frangieh_2021_raw.html#pertpy.data.frangieh_2021_raw" title="pertpy.data.frangieh_2021_raw"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.frangieh_2021_raw</span></code></a>()</p></td>
<td><p>Raw Perturb-CITE-seq data with multi-modal RNA and protein single-cell profiling readout.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.frangieh_2021_rna.html#pertpy.data.frangieh_2021_rna" title="pertpy.data.frangieh_2021_rna"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.frangieh_2021_rna</span></code></a>()</p></td>
<td><p>scPerturb Single-Cell Perturbation Data.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.gasperini_2019_atscale.html#pertpy.data.gasperini_2019_atscale" title="pertpy.data.gasperini_2019_atscale"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.gasperini_2019_atscale</span></code></a>()</p></td>
<td><p>scPerturb Single-Cell Perturbation Data.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.gasperini_2019_highmoi.html#pertpy.data.gasperini_2019_highmoi" title="pertpy.data.gasperini_2019_highmoi"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.gasperini_2019_highmoi</span></code></a>()</p></td>
<td><p>scPerturb Single-Cell Perturbation Data.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.gasperini_2019_lowmoi.html#pertpy.data.gasperini_2019_lowmoi" title="pertpy.data.gasperini_2019_lowmoi"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.gasperini_2019_lowmoi</span></code></a>()</p></td>
<td><p>scPerturb Single-Cell Perturbation Data.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.gehring_2019.html#pertpy.data.gehring_2019" title="pertpy.data.gehring_2019"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.gehring_2019</span></code></a>()</p></td>
<td><p>scPerturb Single-Cell Perturbation Data.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.haber_2017_regions.html#pertpy.data.haber_2017_regions" title="pertpy.data.haber_2017_regions"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.haber_2017_regions</span></code></a>()</p></td>
<td><p>Raw single-cell, pooled CRISPR screening.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.kang_2018.html#pertpy.data.kang_2018" title="pertpy.data.kang_2018"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.kang_2018</span></code></a>()</p></td>
<td><p>Processed multiplexing droplet-based single cell RNA-sequencing using genetic barcodes</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.mcfarland_2020.html#pertpy.data.mcfarland_2020" title="pertpy.data.mcfarland_2020"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.mcfarland_2020</span></code></a>()</p></td>
<td><p>scPerturb Single-Cell Perturbation Data.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.norman_2019.html#pertpy.data.norman_2019" title="pertpy.data.norman_2019"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.norman_2019</span></code></a>()</p></td>
<td><p>Processed single-cell, pooled CRISPR screening.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.norman_2019_filtered.html#pertpy.data.norman_2019_filtered" title="pertpy.data.norman_2019_filtered"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.norman_2019_filtered</span></code></a>()</p></td>
<td><p>scPerturb Single-Cell Perturbation Data.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.norman_2019_raw.html#pertpy.data.norman_2019_raw" title="pertpy.data.norman_2019_raw"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.norman_2019_raw</span></code></a>()</p></td>
<td><p>Raw single-cell, pooled CRISPR screening.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.papalexi_2021.html#pertpy.data.papalexi_2021" title="pertpy.data.papalexi_2021"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.papalexi_2021</span></code></a>()</p></td>
<td><p>Dataset of the Mixscape Vignette.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.papalexi_2021_eccite_arrayed_protein.html#pertpy.data.papalexi_2021_eccite_arrayed_protein" title="pertpy.data.papalexi_2021_eccite_arrayed_protein"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.papalexi_2021_eccite_arrayed_protein</span></code></a>()</p></td>
<td><p>scPerturb Single-Cell Perturbation Data.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.papalexi_2021_eccite_arrayed_rna.html#pertpy.data.papalexi_2021_eccite_arrayed_rna" title="pertpy.data.papalexi_2021_eccite_arrayed_rna"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.papalexi_2021_eccite_arrayed_rna</span></code></a>()</p></td>
<td><p>scPerturb Single-Cell Perturbation Data.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.papalexi_2021_eccite_protein.html#pertpy.data.papalexi_2021_eccite_protein" title="pertpy.data.papalexi_2021_eccite_protein"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.papalexi_2021_eccite_protein</span></code></a>()</p></td>
<td><p>scPerturb Single-Cell Perturbation Data.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.papalexi_2021_eccite_rna.html#pertpy.data.papalexi_2021_eccite_rna" title="pertpy.data.papalexi_2021_eccite_rna"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.papalexi_2021_eccite_rna</span></code></a>()</p></td>
<td><p>scPerturb Single-Cell Perturbation Data.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.replogle_2022_k562_essential.html#pertpy.data.replogle_2022_k562_essential" title="pertpy.data.replogle_2022_k562_essential"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.replogle_2022_k562_essential</span></code></a>()</p></td>
<td><p>scPerturb Single-Cell Perturbation Data.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.replogle_2022_k562_gwps.html#pertpy.data.replogle_2022_k562_gwps" title="pertpy.data.replogle_2022_k562_gwps"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.replogle_2022_k562_gwps</span></code></a>()</p></td>
<td><p>scPerturb Single-Cell Perturbation Data.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.replogle_2022_rpe1.html#pertpy.data.replogle_2022_rpe1" title="pertpy.data.replogle_2022_rpe1"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.replogle_2022_rpe1</span></code></a>()</p></td>
<td><p>scPerturb Single-Cell Perturbation Data.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.sc_sim_augur.html#pertpy.data.sc_sim_augur" title="pertpy.data.sc_sim_augur"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.sc_sim_augur</span></code></a>()</p></td>
<td><p>Simulated test dataset used the Usage example for the Augur.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.schiebinger_2019_16day.html#pertpy.data.schiebinger_2019_16day" title="pertpy.data.schiebinger_2019_16day"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.schiebinger_2019_16day</span></code></a>()</p></td>
<td><p>scPerturb Single-Cell Perturbation Data.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.schiebinger_2019_18day.html#pertpy.data.schiebinger_2019_18day" title="pertpy.data.schiebinger_2019_18day"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.schiebinger_2019_18day</span></code></a>()</p></td>
<td><p>scPerturb Single-Cell Perturbation Data.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.schraivogel_2020_tap_screen_chr8.html#pertpy.data.schraivogel_2020_tap_screen_chr8" title="pertpy.data.schraivogel_2020_tap_screen_chr8"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.schraivogel_2020_tap_screen_chr8</span></code></a>()</p></td>
<td><p>scPerturb Single-Cell Perturbation Data.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.schraivogel_2020_tap_screen_chr11.html#pertpy.data.schraivogel_2020_tap_screen_chr11" title="pertpy.data.schraivogel_2020_tap_screen_chr11"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.schraivogel_2020_tap_screen_chr11</span></code></a>()</p></td>
<td><p>scPerturb Single-Cell Perturbation Data.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.sciplex3_raw.html#pertpy.data.sciplex3_raw" title="pertpy.data.sciplex3_raw"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.sciplex3_raw</span></code></a>()</p></td>
<td><p>Raw sciplex3 perturbation dataset curated for perturbation modeling.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.shifrut_2018.html#pertpy.data.shifrut_2018" title="pertpy.data.shifrut_2018"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.shifrut_2018</span></code></a>()</p></td>
<td><p>scPerturb Single-Cell Perturbation Data.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.smillie.html#pertpy.data.smillie" title="pertpy.data.smillie"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.smillie</span></code></a>()</p></td>
<td><p>scRNA-seq data of the small intestine of mice under Ulcerative Colitis.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.srivatsan_2020_sciplex2.html#pertpy.data.srivatsan_2020_sciplex2" title="pertpy.data.srivatsan_2020_sciplex2"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.srivatsan_2020_sciplex2</span></code></a>()</p></td>
<td><p>scPerturb Single-Cell Perturbation Data.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.srivatsan_2020_sciplex3.html#pertpy.data.srivatsan_2020_sciplex3" title="pertpy.data.srivatsan_2020_sciplex3"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.srivatsan_2020_sciplex3</span></code></a>()</p></td>
<td><p>scPerturb Single-Cell Perturbation Data.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.srivatsan_2020_sciplex4.html#pertpy.data.srivatsan_2020_sciplex4" title="pertpy.data.srivatsan_2020_sciplex4"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.srivatsan_2020_sciplex4</span></code></a>()</p></td>
<td><p>scPerturb Single-Cell Perturbation Data.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.stephenson_2021_subsampled.html#pertpy.data.stephenson_2021_subsampled" title="pertpy.data.stephenson_2021_subsampled"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.stephenson_2021_subsampled</span></code></a>()</p></td>
<td><p>Processed 10X 5' scRNA-seq data from PBMC of COVID-19 patients and healthy donors</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.tian_2019_day7neuron.html#pertpy.data.tian_2019_day7neuron" title="pertpy.data.tian_2019_day7neuron"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.tian_2019_day7neuron</span></code></a>()</p></td>
<td><p>scPerturb Single-Cell Perturbation Data.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.tian_2019_ipsc.html#pertpy.data.tian_2019_ipsc" title="pertpy.data.tian_2019_ipsc"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.tian_2019_ipsc</span></code></a>()</p></td>
<td><p>scPerturb Single-Cell Perturbation Data.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.tian_2021_crispra.html#pertpy.data.tian_2021_crispra" title="pertpy.data.tian_2021_crispra"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.tian_2021_crispra</span></code></a>()</p></td>
<td><p>scPerturb Single-Cell Perturbation Data.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.tian_2021_crispri.html#pertpy.data.tian_2021_crispri" title="pertpy.data.tian_2021_crispri"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.tian_2021_crispri</span></code></a>()</p></td>
<td><p>scPerturb Single-Cell Perturbation Data.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.weinreb_2020.html#pertpy.data.weinreb_2020" title="pertpy.data.weinreb_2020"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.weinreb_2020</span></code></a>()</p></td>
<td><p>scPerturb Single-Cell Perturbation Data.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="data/pertpy.data.xie_2017.html#pertpy.data.xie_2017" title="pertpy.data.xie_2017"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.xie_2017</span></code></a>()</p></td>
<td><p>scPerturb Single-Cell Perturbation Data.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="data/pertpy.data.zhao_2021.html#pertpy.data.zhao_2021" title="pertpy.data.zhao_2021"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data.zhao_2021</span></code></a>()</p></td>
<td><p>scPerturb Single-Cell Perturbation Data.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="tools">
<h2>Tools<a class="headerlink" href="#tools" title="Permalink to this heading">#</a></h2>
<section id="augurpy">
<h3>Augurpy<a class="headerlink" href="#augurpy" title="Permalink to this heading">#</a></h3>
<p>The Python implementation of <a class="reference external" href="https://github.com/neurorestore/Augur">Augur R package</a> Skinnider, M.A., Squair, J.W., Kathe, C. et al. <a class="reference external" href="https://doi.org/10.1038/s41587-020-0605-1">Cell type prioritization in single-cell data</a>. Nat Biotechnol 39, 30â€“34 (2021).</p>
<p>Augurpy aims to rank or prioritize cell types according to their response to experimental perturbations given high dimensional single-cell sequencing data.
The basic idea is that in the space of molecular measurements cells reacting heavily to induced perturbations are
more easily separated into perturbed and unperturbed than cell types with little or no response.
This separability is quantified by measuring how well experimental labels (eg. treatment and control) can be predicted within each cell type.
Augurpy trains a machine learning model predicting experimental labels for each cell type in multiple cross validation runs and
then prioritizes cell type response according to metric scores measuring the accuracy of the model.
For categorical data the area under the curve is the default metric and for numerical data the concordance correlation coefficient
is used as a proxy for how accurate the model is which in turn approximates perturbation response.</p>
<p>Example implementation:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pertpy</span> <span class="k">as</span> <span class="nn">pt</span>

<span class="n">adata</span> <span class="o">=</span> <span class="n">pt</span><span class="o">.</span><span class="n">dt</span><span class="o">.</span><span class="n">sc_sim_augur</span><span class="p">()</span>
<span class="n">ag</span> <span class="o">=</span> <span class="n">pt</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">Augurpy</span><span class="p">(</span><span class="n">estimator</span><span class="o">=</span><span class="s2">&quot;random_forest_classifier&quot;</span><span class="p">)</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">ag</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">adata</span><span class="p">,</span> <span class="n">results</span> <span class="o">=</span> <span class="n">ag</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>

<span class="c1"># metrics for each cell type</span>
<span class="n">results</span><span class="p">[</span><span class="s1">&#39;summary_metrics&#39;</span><span class="p">]</span>
</pre></div>
</div>
<p>See <a class="reference external" href="https://pertpy.readthedocs.io/en/latest/tutorials/notebooks/augurpy.html">augurpy tutorial</a> for a more elaborate tutorial.</p>
<div class="table-wrapper autosummary longtable docutils container">
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="tools/pertpy.tools.Augurpy.html#pertpy.tools.Augurpy" title="pertpy.tools.Augurpy"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tools.Augurpy</span></code></a></p></td>
<td><p>Python implementation of Augur.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="mixscape">
<h3>Mixscape<a class="headerlink" href="#mixscape" title="Permalink to this heading">#</a></h3>
<p>A Python implementation of <a class="reference external" href="https://satijalab.org/seurat/articles/mixscape_vignette.html">Mixscape</a> Papalexi et al. <a class="reference external" href="https://www.nature.com/articles/s41588-021-00778-2">Characterizing the molecular regulation of inhibitory immune checkpoints with multimodal single-cell screens</a>.</p>
<p>Mixscape first tries to remove confounding sources of variation such as cell cycle or replicate effect by embedding the cells into a perturbation space (the perturbation signature).
Next, it determines which targeted cells were affected by the genetic perturbation (=KO) and which targeted cells were not (=NP) with the use of mixture models.
Finally, it visualizes similarities and differences across different perturbations.</p>
<div class="table-wrapper autosummary longtable docutils container">
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="tools/pertpy.tools.Mixscape.html#pertpy.tools.Mixscape" title="pertpy.tools.Mixscape"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tools.Mixscape</span></code></a>()</p></td>
<td><p>Python implementation of Mixscape.</p></td>
</tr>
</tbody>
</table>
</div>
<p>Example implementation:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pertpy</span> <span class="k">as</span> <span class="nn">pt</span>

<span class="n">mdata</span> <span class="o">=</span> <span class="n">pt</span><span class="o">.</span><span class="n">dt</span><span class="o">.</span><span class="n">papalexi_2021</span><span class="p">()</span>
<span class="n">ms</span> <span class="o">=</span> <span class="n">pt</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">Mixscape</span><span class="p">()</span>
<span class="n">ms</span><span class="o">.</span><span class="n">pert_sign</span><span class="p">(</span><span class="n">mdata</span><span class="p">[</span><span class="s1">&#39;rna&#39;</span><span class="p">],</span> <span class="s1">&#39;perturbation&#39;</span><span class="p">,</span> <span class="s1">&#39;NT&#39;</span><span class="p">,</span> <span class="s1">&#39;replicate&#39;</span><span class="p">)</span>
<span class="n">ms</span><span class="o">.</span><span class="n">mixscape</span><span class="p">(</span><span class="n">adata</span> <span class="o">=</span> <span class="n">mdata</span><span class="p">[</span><span class="s1">&#39;rna&#39;</span><span class="p">],</span> <span class="n">control</span> <span class="o">=</span> <span class="s1">&#39;NT&#39;</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="s1">&#39;gene_target&#39;</span><span class="p">,</span> <span class="n">layer</span><span class="o">=</span><span class="s1">&#39;X_pert&#39;</span><span class="p">)</span>
<span class="n">ms</span><span class="o">.</span><span class="n">lda</span><span class="p">(</span><span class="n">adata</span><span class="o">=</span><span class="n">mdata</span><span class="p">[</span><span class="s1">&#39;rna&#39;</span><span class="p">],</span> <span class="n">labels</span><span class="o">=</span><span class="s1">&#39;gene_target&#39;</span><span class="p">,</span> <span class="n">layer</span><span class="o">=</span><span class="s1">&#39;X_pert&#39;</span><span class="p">)</span>
<span class="n">pt</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">ms</span><span class="o">.</span><span class="n">lda</span><span class="p">(</span><span class="n">adata</span><span class="o">=</span><span class="n">mdata</span><span class="p">[</span><span class="s1">&#39;rna&#39;</span><span class="p">])</span>
</pre></div>
</div>
<p>See <a class="reference external" href="https://pertpy.readthedocs.io/en/latest/tutorials/notebooks/mixscape.html">mixscape tutorial</a> for a more elaborate tutorial.</p>
</section>
<section id="compositional-analysis">
<h3>Compositional analysis<a class="headerlink" href="#compositional-analysis" title="Permalink to this heading">#</a></h3>
<section id="milopy">
<h4>Milopy<a class="headerlink" href="#milopy" title="Permalink to this heading">#</a></h4>
<p>A Python implementation of Milo for differential abundance testing on KNN graphs, to ease interoperability with scverse pipelines for single-cell analysis.
See <a class="reference external" href="https://www.nature.com/articles/s41587-021-01033-z">Differential abundance testing on single-cell data using k-nearest neighbor graphs</a> for details on the statistical framework.</p>
<div class="table-wrapper autosummary longtable docutils container">
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="tools/pertpy.tools.Milopy.html#pertpy.tools.Milopy" title="pertpy.tools.Milopy"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tools.Milopy</span></code></a>()</p></td>
<td><p>Python implementation of Milo.</p></td>
</tr>
</tbody>
</table>
</div>
<p>See <a class="reference external" href="https://pertpy.readthedocs.io/en/latest/tutorials/notebooks/milopy.html">milopy tutorial</a> for a more elaborate tutorial.</p>
</section>
<section id="sccoda-and-tasccoda">
<h4>scCODA and tascCODA<a class="headerlink" href="#sccoda-and-tasccoda" title="Permalink to this heading">#</a></h4>
<p>Reimplementation of scCODA for identification of compositional changes in high-throughput sequencing count data and tascCODA for sparse, tree-aggregated modeling of high-throughput sequencing data.
See <a class="reference external" href="https://www.nature.com/articles/s41467-021-27150-6">scCODA is a Bayesian model for compositional single-cell data analysis</a> for statistical methodology and benchmarking performance of scCODA and <a class="reference external" href="https://www.frontiersin.org/articles/10.3389/fgene.2021.766405/full">tascCODA: Bayesian Tree-Aggregated Analysis of Compositional Amplicon and Single-Cell Data</a> for statistical methodology and benchmarking performance of tascCODA.</p>
<div class="table-wrapper autosummary longtable docutils container">
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="tools/pertpy.tools.Sccoda.html#pertpy.tools.Sccoda" title="pertpy.tools.Sccoda"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tools.Sccoda</span></code></a>(*args, **kwargs)</p></td>
<td><p>Statistical model for single-cell differential composition analysis with specification of a reference cell type.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="tools/pertpy.tools.Tasccoda.html#pertpy.tools.Tasccoda" title="pertpy.tools.Tasccoda"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tools.Tasccoda</span></code></a>(*args, **kwargs)</p></td>
<td><p>Statistical model for tree-aggregated differential composition analysis (tascCODA, Ostner et al., 2021).</p></td>
</tr>
</tbody>
</table>
</div>
</section>
</section>
<section id="representation">
<h3>Representation<a class="headerlink" href="#representation" title="Permalink to this heading">#</a></h3>
<div class="table-wrapper autosummary longtable docutils container">
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="tools/pertpy.tools.kernel_pca.html#pertpy.tools.kernel_pca" title="pertpy.tools.kernel_pca"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tools.kernel_pca</span></code></a>(adata[, n_comps, kernel, ...])</p></td>
<td><p>Compute kernel PCA on <cite>adata.X</cite>.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
</section>
<section id="plots">
<h2>Plots<a class="headerlink" href="#plots" title="Permalink to this heading">#</a></h2>
<section id="id1">
<h3>Augurpy<a class="headerlink" href="#id1" title="Permalink to this heading">#</a></h3>
<div class="table-wrapper autosummary longtable docutils container">
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="plot/pertpy.plot.ag.dp_scatter.html#pertpy.plot.ag.dp_scatter" title="pertpy.plot.ag.dp_scatter"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot.ag.dp_scatter</span></code></a>(results[, top_n, ax, ...])</p></td>
<td><p>Plot result of differential prioritization.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="plot/pertpy.plot.ag.important_features.html#pertpy.plot.ag.important_features" title="pertpy.plot.ag.important_features"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot.ag.important_features</span></code></a>(data[, key, ...])</p></td>
<td><p>Plot a lollipop plot of the n features with largest feature importances.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="plot/pertpy.plot.ag.lollipop.html#pertpy.plot.ag.lollipop" title="pertpy.plot.ag.lollipop"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot.ag.lollipop</span></code></a>(data[, key, ax, return_figure])</p></td>
<td><p>Plot a lollipop plot of the mean augur values.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="plot/pertpy.plot.ag.scatterplot.html#pertpy.plot.ag.scatterplot" title="pertpy.plot.ag.scatterplot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot.ag.scatterplot</span></code></a>(results1, results2[, ...])</p></td>
<td><p>Create scatterplot with two augur results.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="id2">
<h3>Mixscape<a class="headerlink" href="#id2" title="Permalink to this heading">#</a></h3>
<div class="table-wrapper autosummary longtable docutils container">
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="plot/pertpy.plot.ms.violin.html#pertpy.plot.ms.violin" title="pertpy.plot.ms.violin"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot.ms.violin</span></code></a>(adata, target_gene_idents[, ...])</p></td>
<td><p>Violin plot using mixscape results.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="plot/pertpy.plot.ms.perturbscore.html#pertpy.plot.ms.perturbscore" title="pertpy.plot.ms.perturbscore"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot.ms.perturbscore</span></code></a>(adata, labels, target_gene)</p></td>
<td><p>Density plots to visualize perturbation scores calculated by the <cite>pt.tl.mixscape</cite> function.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="plot/pertpy.plot.ms.heatmap.html#pertpy.plot.ms.heatmap" title="pertpy.plot.ms.heatmap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot.ms.heatmap</span></code></a>(adata, labels, target_gene)</p></td>
<td><p>Heatmap plot using mixscape results.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="plot/pertpy.plot.ms.barplot.html#pertpy.plot.ms.barplot" title="pertpy.plot.ms.barplot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot.ms.barplot</span></code></a>(adata[, control, ...])</p></td>
<td><p>Barplot to visualize perturbation scores calculated from RunMixscape function.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="plot/pertpy.plot.ms.lda.html#pertpy.plot.ms.lda" title="pertpy.plot.ms.lda"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot.ms.lda</span></code></a>(adata[, mixscape_class, ...])</p></td>
<td><p>Visualizing perturbation responses with Linear Discriminant Analysis.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="id3">
<h3>Compositional analysis<a class="headerlink" href="#id3" title="Permalink to this heading">#</a></h3>
<section id="id4">
<h4>Milopy<a class="headerlink" href="#id4" title="Permalink to this heading">#</a></h4>
<div class="table-wrapper autosummary longtable docutils container">
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="plot/pertpy.plot.milo.nhood_graph.html#pertpy.plot.milo.nhood_graph" title="pertpy.plot.milo.nhood_graph"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot.milo.nhood_graph</span></code></a>(mdata[, alpha, ...])</p></td>
<td><p>Visualize DA results on abstracted graph (wrapper around sc.pl.embedding)</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="plot/pertpy.plot.milo.nhood.html#pertpy.plot.milo.nhood" title="pertpy.plot.milo.nhood"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot.milo.nhood</span></code></a>(mdata, ix[, feature_key, ...])</p></td>
<td><p>Visualize cells in a neighbourhood.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="plot/pertpy.plot.milo.da_beeswarm.html#pertpy.plot.milo.da_beeswarm" title="pertpy.plot.milo.da_beeswarm"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot.milo.da_beeswarm</span></code></a>(mdata[, feature_key, ...])</p></td>
<td><p>Plot beeswarm plot of logFC against nhood labels</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="plot/pertpy.plot.milo.nhood_counts_by_cond.html#pertpy.plot.milo.nhood_counts_by_cond" title="pertpy.plot.milo.nhood_counts_by_cond"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot.milo.nhood_counts_by_cond</span></code></a>(mdata, test_var)</p></td>
<td><p>Plot boxplot of cell numbers vs condition of interest</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="id5">
<h4>scCODA and tascCODA<a class="headerlink" href="#id5" title="Permalink to this heading">#</a></h4>
<div class="table-wrapper autosummary longtable docutils container">
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="plot/pertpy.plot.coda.stacked_barplot.html#pertpy.plot.coda.stacked_barplot" title="pertpy.plot.coda.stacked_barplot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot.coda.stacked_barplot</span></code></a>(data, feature_name)</p></td>
<td><p>Plots a stacked barplot for all levels of a covariate or all samples (if feature_name==&quot;samples&quot;).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="plot/pertpy.plot.coda.effects_barplot.html#pertpy.plot.coda.effects_barplot" title="pertpy.plot.coda.effects_barplot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot.coda.effects_barplot</span></code></a>(data[, ...])</p></td>
<td><p>Barplot visualization for effects.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="plot/pertpy.plot.coda.boxplots.html#pertpy.plot.coda.boxplots" title="pertpy.plot.coda.boxplots"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot.coda.boxplots</span></code></a>(data, feature_name[, ...])</p></td>
<td><p>Grouped boxplot visualization. The cell counts for each cell type are shown as a group of boxplots,</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="plot/pertpy.plot.coda.rel_abundance_dispersion_plot.html#pertpy.plot.coda.rel_abundance_dispersion_plot" title="pertpy.plot.coda.rel_abundance_dispersion_plot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot.coda.rel_abundance_dispersion_plot</span></code></a>(data)</p></td>
<td><p>Plots total variance of relative abundance versus minimum relative abundance of all cell types for determination of a reference cell type.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="plot/pertpy.plot.coda.draw_tree.html#pertpy.plot.coda.draw_tree" title="pertpy.plot.coda.draw_tree"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot.coda.draw_tree</span></code></a>(data[, modality_key, ...])</p></td>
<td><p>Plot a tree using input ete3 tree object.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="plot/pertpy.plot.coda.draw_effects.html#pertpy.plot.coda.draw_effects" title="pertpy.plot.coda.draw_effects"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot.coda.draw_effects</span></code></a>(data, covariate[, ...])</p></td>
<td><p>Plot a tree with colored circles on the nodes indicating significant effects with bar plots which indicate leave-level significant effects.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="plot/pertpy.plot.coda.effects_umap.html#pertpy.plot.coda.effects_umap" title="pertpy.plot.coda.effects_umap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot.coda.effects_umap</span></code></a>(data, effect_name, ...)</p></td>
<td><p>Plot a UMAP visualization colored by effect strength.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
</section>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="data/pertpy.data.adamson_2016_pilot.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">pertpy.data.adamson_2016_pilot</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="../installation.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Installation</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2021, Lukas Heumos, Theislab
            </div>
            Made with 
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">Usage</a><ul>
<li><a class="reference internal" href="#data">Data</a></li>
<li><a class="reference internal" href="#tools">Tools</a><ul>
<li><a class="reference internal" href="#augurpy">Augurpy</a></li>
<li><a class="reference internal" href="#mixscape">Mixscape</a></li>
<li><a class="reference internal" href="#compositional-analysis">Compositional analysis</a><ul>
<li><a class="reference internal" href="#milopy">Milopy</a></li>
<li><a class="reference internal" href="#sccoda-and-tasccoda">scCODA and tascCODA</a></li>
</ul>
</li>
<li><a class="reference internal" href="#representation">Representation</a></li>
</ul>
</li>
<li><a class="reference internal" href="#plots">Plots</a><ul>
<li><a class="reference internal" href="#id1">Augurpy</a></li>
<li><a class="reference internal" href="#id2">Mixscape</a></li>
<li><a class="reference internal" href="#id3">Compositional analysis</a><ul>
<li><a class="reference internal" href="#id4">Milopy</a></li>
<li><a class="reference internal" href="#id5">scCODA and tascCODA</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/sphinx_highlight.js"></script>
    <script src="../_static/scripts/furo.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script src="../_static/design-tabs.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    </body>
</html>